<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title>Bismark User Guide</title>
  <style type="text/css">
    body{font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";margin:10px;color:#333}
    blockquote{font-style:italic;font-size:.9em;color:#999}
    h1{margin-top:80px;padding-bottom:5px;border-bottom:1px solid #dedede}
    img{max-width:100%}
    pre{max-width:100%;overflow:auto;background-color:#ededed;border-radius:5px;padding:8px}
    code{white-space:pre-wrap;background-color:#f9f9f9;padding:3px 5px;font-family:Consolas,"Liberation Mono",Menlo,Courier,monospace;font-size:90%}
    pre code{background-color:#ededed;white-space:pre}
    hr{margin:30px 0;border:0;border-top:1px solid #dedede;height:0}
    table { border-collapse: collapse; }
    table tbody tr { border-top: 1px solid #ccc; }
    table tbody tr td, table thead tr th { padding: 6px 13px; border: 1px solid #ddd }
    table tbody tr:nth-child(2n) { background-color: #f8f8f8; }
    #header_img{float:right;max-width:30%;margin-top:-60px}
    #TOC::before{content:'Table of Contents';font-weight:700;font-size:2em}
    #TOC{font-size:.9em;background-color:#ededed;border-radius:10px;padding:10px}
    #TOC ul{padding:0;list-style-type:none;margin-bottom:10px}
    #TOC ul ul{margin-left:10px;text-decoration:italic}
    #TOC ul li{margin-top:5px}
    #TOC a{text-decoration:none;color:#333}
    #TOC ul ul a{color:#999}
    #TOC ul ul a code{background-color:transparent}
    @media screen and (min-width: 700px) {
      body{margin:20px 420px 20px 20px}
      #TOC{position:fixed;right:20px;top:20px;bottom:20px;width:350px;max-height:100%;overflow:auto}
    }
    #bismark-bisulfite-mapper{margin-top:0;font-size:6em;font-weight:100;border:none}
  </style>
  <!--[if lt IE 9]><script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script><![endif]-->
</head>
<body>
<nav id="TOC"><ul><li><a href="#bismark-bisulfite-mapper">Bismark Bisulfite Mapper</a><ul><li><a href="#user-guide-v">User Guide - v0.21.0</a></ul></li><li><a href="#quick-reference">1) Quick Reference</a><ul><li><a href="#i-running-bismark-genome-preparation">(I) Running <code>bismark_genome_preparation</code></a></li><li><a href="#ii-running-bismark">(II) Running <code>bismark</code></a></li><li><a href="#iii-running-deduplicate-bismark">(III) Running <code>deduplicate_bismark</code></a></li><li><a href="#iv-running-bismark-methylation-extractor">(IV) Running <code>bismark_methylation_extractor</code></a></li><li><a href="#v-running-bismark-report">(V) Running <code>bismark2report</code></a></li><li><a href="#vi-running-bismark-summary">(VI) Running <code>bismark2summary</code></a></ul></li><li><a href="#bismark-general-information">2) Bismark - General Information</a><ul><li><a href="#what-is-bismark">What is Bismark?</a></li><li><a href="#installation-notes">Installation notes</a></li><li><a href="#dependencies">Dependencies</a></li><li><a href="#hardware-requirements">Hardware requirements</a></li><li><a href="#bs-seq-test-data-set">BS-Seq test data set</a></li><li><a href="#which-kind-of-bs-seq-files-are-supported">Which kind of BS-Seq files are supported?</a></li><li><a href="#how-does-bismark-work">How does Bismark work?</a></li><li><a href="#bismark-alignment-and-methylation-call-report">Bismark alignment and methylation call report</a></ul></li><li><a href="#running-bismark">3) Running Bismark</a><ul><li><a href="#i-bismark-genome-preparation">(I) Bismark Genome Preparation</a></li><li><a href="#ii-bismark-alignment-step">(II) Bismark Alignment Step</a></li><li><a href="#iii-bismark-deduplication-step">(III) Bismark Deduplication Step</a></li><li><a href="#iv-bismark-methylation-extractor">(IV) Bismark methylation extractor</a></li><li><a href="#v-the-bismark-html-processing-report">(V) The Bismark HTML Processing Report</a></li><li><a href="#vi-the-bismark-summary-report">(VI) The Bismark Summary Report</a></li><li><a href="#vii-bismark-nucleotide-coverage-report-bam-nuc">(VII) Bismark Nucleotide Coverage report (<code>bam2nuc</code>)</a></li><li><a href="#viii-filtering-out-non-bisulfite-converted-reads-filter-non-conversion">(VIII) Filtering out non-bisulfite converted reads (<code>filter_non_conversion</code>)</a></li><li><a href="#ix-notes-about-different-library-types-and-commercial-kits">(IX) Notes about different library types and commercial kits</a></ul></li><li><a href="#appendix-full-list-of-options">4) APPENDIX - Full list of options</a><ul><li><a href="#appendix-i-bismark-genome-preparation">Appendix (I): Bismark Genome Preparation</a></li><li><a href="#appendix-ii-bismark">Appendix (II): Bismark</a></li><li><a href="#appendix-iii-bismark-methylation-extractor">Appendix (III): Bismark Methylation Extractor</a></li><li><a href="#appendix-iv-bismark-reports-for-the-test-data-set">Appendix (IV): Bismark reports for the test data set</a></ul></li><li><a href="#credits">Credits</a></li></ul></nav>
<h1 id="bismark-bisulfite-mapper">Bismark Bisulfite Mapper</h1>

<p><img title="Bismark" id="header_img" src="Images/bismark.png"></p>

<h2 id="user-guide-v">User Guide - v0.21.0</h2>

<h4 id="february">26 February, 2019</h4>

<p>This User Guide outlines the Bismark suite of tools and gives more details for each individual step. For troubleshooting some of the more commonly experienced problems in sequencing in general and bisulfite-sequencing in particular please browse through the sequencing section at <a href="https://sequencing.qcfail.com/">QCFail.com</a>.</p>

<h1 id="quick-reference">1) Quick Reference</h1>

<p>Bismark needs a working version of Perl and it is run from the command line. Furthermore, <a href="http://bowtie-bio.sourceforge.net/bowtie2">Bowtie 2</a> or <a href="https://ccb.jhu.edu/software/hisat2/index.shtml">HISAT2</a> needs to be installed on your computer. For more information on how to run Bismark with Bowtie 2 please go to the end of this manual.</p>

<p>As of version 0.14.0 or higher, Bismark may be run using parallelisation for both the alignment and the methylation extraction step. Search for <code>--parallel/--multicore</code> for more details below.</p>

<p>First you need to download a reference genome and place it in a genome folder. Genomes can be obtained e.g. from the <a href="http://www.ensembl.org/info/data/ftp/index.html/">Ensembl</a> or <a href="ftp://ftp.ncbi.nih.gov/genomes/">NCBI</a> websites. For the example below you would need to download the <em>Homo sapiens</em> genome. Bismark supports reference genome sequence files in <code>FastA</code> format, allowed file extensions are either either <code>.fa</code> or <code>.fasta</code>. Both single-entry and multiple-entry <code>FastA</code> files are supported.</p>

<p>The following examples will use the file <code>test_dataset.fastq</code> which is available for download from the Bismark project or Github pages (it contains 10,000 reads in FastQ format, Phred33 qualities, 50 bp long reads, from a human directional BS-Seq library). An example report can be found in Appendix IV.</p>

<h2 id="i-running-bismark-genome-preparation">(I) Running <code>bismark_genome_preparation</code></h2>

<p><strong>USAGE:</strong>
<code>
bismark_genome_preparation [options] &lt;path_to_genome_folder&gt;
</code></p>

<p>A typical genome indexing could look like this:
<code>
/bismark/bismark_genome_preparation --path_to_aligner /usr/bin/bowtie2/ --verbose /data/genomes/homo_sapiens/GRCh37/
</code></p>

<h2 id="ii-running-bismark">(II) Running <code>bismark</code></h2>

<p><strong>USAGE:</strong>
<code>
bismark [options] --genome &lt;genome_folder&gt; {-1 &lt;mates1&gt; -2 &lt;mates2&gt; | &lt;singles&gt;}
</code></p>

<p>Typical alignment example:
<code>
bismark --genome /data/genomes/homo_sapiens/GRCh37/ test_dataset.fastq
</code></p>

<p>This will produce two output files:
1. <code>test_dataset_bismark_bt2.bam</code> (contains all alignments plus methylation call strings)
2. <code>test_dataset_bismark_SE_report.txt</code> (contains alignment and methylation summary)</p>

<p><strong>NOTE:</strong> In order to work properly the current working directory must contain the sequence files to be analysed.</p>

<h2 id="iii-running-deduplicate-bismark">(III) Running <code>deduplicate_bismark</code></h2>

<p><code>
deduplicate_bismark --bam [options] &lt;filenames&gt;
</code></p>

<p>This command will deduplicate the Bismark alignment BAM file and remove all reads but one which align to the the very same position and in the same orientation. This step is recommended for whole-genome bisulfite samples, but should not be used for reduced representation libraries such as RRBS, amplicon or target enrichment libraries.</p>

<h2 id="iv-running-bismark-methylation-extractor">(IV) Running <code>bismark_methylation_extractor</code></h2>

<p><strong>USAGE:</strong>
<code>
bismark_methylation_extractor [options] &lt;filenames&gt;
</code></p>

<p>A typical command to extract context-dependent (CpG/CHG/CHH) methylation could look like this:
<code>
bismark_methylation_extractor --gzip --bedGraph test_dataset_bismark_bt2.bam
</code></p>

<p>This will produce three methytlation output files:</p>

<ul>
<li><code>CpG_context_test_dataset_bismark_bt2.txt.gz</code>
</li>
<li><code>CHG_context_test_dataset_bismark_bt2.txt.gz</code>
</li>
<li><code>CHH_context_test_dataset_bismark_bt2.txt.gz</code>
</li>
</ul>

<p>as well as a bedGraph and a Bismark coverage file. For more on these files and their formats please see below.</p>

<h2 id="v-running-bismark-report">(V) Running <code>bismark2report</code></h2>

<p><strong>USAGE:</strong>
<code>
bismark2report [options]
</code></p>

<p>This command attempts to find Bismark alignment, deduplication and methylation extraction (splitting) reports as well as M-bias files to generate a graphical HTML report such as this <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/PE_report.html">example Bismark paired-end report</a> for each sample in a directory.</p>

<h2 id="vi-running-bismark-summary">(VI) Running <code>bismark2summary</code></h2>

<p><strong>USAGE:</strong>
<code>
bismark2summary [options]
</code></p>

<p>This command scans the current working directory for different Bismark alignment, deduplication and methylation extraction (splitting) reports to produce a graphical summary HTML report, as well as a data table, for all files in a directory. Here is a sample <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/bismark_summary_WGBS.html">Bismark Summary Report</a>.</p>

<h1 id="bismark-general-information">2) Bismark - General Information</h1>

<h2 id="what-is-bismark">What is Bismark?</h2>

<p>Bismark is a set of tools for the time-efficient analysis of Bisulfite-Seq (BS-Seq) data. Bismark performs alignments of bisulfite-treated reads to a reference genome and cytosine methylation calls at the same time. Bismark is written in Perl and is run from the command line. Bisulfite-treated reads are mapped using the short read aligner Bowtie 2, or alternatively HISAT2. Therefore, it is a requirement that Bowtie 2 (or HISAT2) are also installed on your machine (see Dependencies).</p>

<p>All files associated with Bismark as well as a test BS-Seq data set can be downloaded from <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Babraham Bioinformatics</a> or <a href="https://github.com/FelixKrueger/Bismark">Github</a>.</p>

<p>We would like to hear your comments, suggestions or bugs about Bismark! Please e-mail them to: <a href="mailto:felix.krueger@babraham.ac.uk">felix.krueger@babraham.ac.uk</a></p>

<h2 id="installation-notes">Installation notes</h2>

<p>Bismark is written in Perl and is executed from the command line. To install Bismark simply copy the bismark_v0.X.Y.tar.gz file into a Bismark installation folder and extract all files by typing:</p>

<p><code>
tar xzf bismark_v0.X.Y.tar.gz
</code></p>

<h2 id="dependencies">Dependencies</h2>

<p>Bismark requires a working of Perl and <a href="http://bowtie-bio.sourceforge.net/bowtie2">Bowtie 2</a> (or <a href="https://ccb.jhu.edu/software/hisat2/index.shtml">HISAT2</a>) to be installed on your machine. Bismark will assume that the Bowtie 2/ HISAT2 executable is in your path unless the path to Bowtie/ HISAT2 is specified manually with:
<code>
--path_to_bowtie2 &lt;/../../bowtie2&gt; or 
--path_to_hisat2 &lt;/../../hisat2&gt;
</code></p>

<h2 id="hardware-requirements">Hardware requirements</h2>

<p>Bismark holds the reference genome in memory, and in addition to that runs up to four parallel instances of Bowtie 2. The memory usage is dependent on the size of the reference genome. For a large eukaryotic genome (human or mouse) we experienced a typical memory usage of around 12GB. We thus recommend running Bismark on a machine with 5 CPU cores and at least 12 GB of RAM. The memory requirements of Bowtie 2 are somewhat larger (possibly to allow gapped alignments). When running Bismark using Bowtie 2 we therefore recommend a system with at least 5 cores and > 16GB of RAM.</p>

<p>Alignment speed depends largely on the read length and alignment parameters used. Allowing many mismatches and using a short seed length tends to be fairly slow.</p>

<h2 id="bs-seq-test-data-set">BS-Seq test data set</h2>

<p>A test BS-Seq data set is available for download from the Bismark project or Github pages. It contains 10,000 single- end shotgun BS reads from human ES cells in FastQ format (from SRR020138, Lister et al., 2009; trimmed to 50 bp; base call qualities are Sanger encoded Phred values (Phred33)).</p>

<h2 id="which-kind-of-bs-seq-files-are-supported">Which kind of BS-Seq files are supported?</h2>

<p>Bismark supports the alignment of bisulfite-treated reads (whole genome shotgun BS-Seq (WGSBS), reduced-representation BS-Seq (RRBS) or PBAT-Seq (Post-Bisulfite Adapter Tagging) for the following conditions:</p>

<ul>
<li>sequence format either <code>FastQ</code> or <code>FastA</code>
</li>
<li>single-end or paired-end reads
</li>
<li>input files can be uncompressed or <code>gzip</code>-compressed (ending in <code>.gz</code>)
</li>
<li>variable read length support
</li>
<li>directional or non-directional BS-Seq libraries
</li>
</ul>

<p>A full list of alignments modes can be found in <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_alignment_modes.pdf"><code>Bismark_alignment_modes.pdf</code></a>.</p>

<p>In addition, Bismark retains much of the flexibility of Bowtie 2 / HISAT2 (adjustable seed length, number of mismatches, insert size ...). For a full list of options please run:</p>

<p><code>
bismark --help
</code></p>

<p>or see the Appendix at the end of this User Guide.</p>

<p><strong>NOTE:</strong> It should be mentioned that Bismark supports only reads in base-space, such as from the Illumina platform. There are currently no plans to extend its functionality to colour-space reads from the SOLiD platform. </p>

<h2 id="how-does-bismark-work">How does Bismark work?</h2>

<p>Sequence reads are first transformed into fully bisulfite-converted forward (C->T) and reverse read (G->A conversion of the forward strand) versions, before they are aligned to similarly converted versions of the genome (also C->T and G->A converted). Sequence reads that produce a unique best alignment from the four alignment processes against the bisulfite genomes (which are running in parallel) are then compared to the normal genomic sequence and the methylation state of all cytosine positions in the read is inferred. A read is considered to align uniquely if an alignment has a unique best alignment score (as reported by the <code>AS:i</code> field). If a read produces several alignments with the same number of mismatches or with the same alignment score (<code>AS:i</code> field), a read (or a read-pair) is discarded altogether.</p>

<h2 id="bismark-alignment-and-methylation-call-report">Bismark alignment and methylation call report</h2>

<p>Upon completion, Bismark produces a run report containing information about the following:
- Summary of alignment parameters used
- Number of sequences analysed
- Number of sequences with a unique best alignment (mapping efficiency)
- Statistics summarising the bisulfite strand the unique best alignments came from
- Number of cytosines analysed
- Number of methylated and unmethylated cytosines
- Percentage methylation of cytosines in CpG, CHG or CHH context (where H can be either A, T or C). This percentage is calculated individually for each context following the equation:</p>

<blockquote>
  <p>% methylation (context) = 100 * methylated Cs (context) / (methylated Cs (context) + unmethylated Cs (context)).</p>
</blockquote>

<p>It should be stressed that the percent methylation value (context) is just a very rough calculation performed directly at the mapping step. Actual methylation levels after post-processing or filtering have been applied may vary.</p>

<h1 id="running-bismark">3) Running Bismark</h1>

<p>Running Bismark is split up into three main steps:</p>

<ol>
<li>First, the genome of interest needs to be bisulfite converted and indexed to allow Bowtie alignments. This step needs to be carried out only once for each genome. Note that Bowtie 2 and HISAT2 require distinct indexing steps since their indexes are not compatible.
</li>
<li>Bismark read alignment step. Simply specify a file to be analysed, a reference genome and alignment parameters. Bismark will produce a combined alignment/methylation call output (default is BAM format) as well as a run statistics report.
</li>
<li>Bismark methylation extractor. This step is optional and will extract the methylation information from the Bismark alignment output. Running this additional step allows splitting the methylation information up into the different contexts, getting strand-specific methylation information and offers some filtering options. You can also choose to sort the methylation information into <code>bedGraph</code>/<code>coverage</code> files, or even process them further to genome-wide cytosine methylation reports.
</li>
</ol>

<p>Each of these steps will be described in more detail (with examples) in the following sections.</p>

<h2 id="i-bismark-genome-preparation">(I) Bismark Genome Preparation</h2>

<p>This script needs to be run only once to prepare the genome of interest for bisulfite alignments. You need to specify a directory containing the genome you want to align your reads against (please be aware that the <code>bismark_genome_preparation</code> script expects <code>FastA</code> files in this folder (with either <code>.fa</code> or <code>.fasta</code> extension, single or multiple sequence entries per file). Bismark will create two individual folders within this directory, one for a C->T converted genome and the other one for the G->A converted genome. After creating C->T and G->A versions of the genome they will be indexed in parallel using the indexer <code>bowtie2-build</code> (or <code>hisat2-build</code>). Once both C->T and G->A genome indices have been created you do not need to use the genome preparation script again (unless you want to align against a different genome...).</p>

<p>Again, <strong>please note</strong> that Bowtie 2 and HISAT2 indexes are not compatible! To create a genome index for use with HISAT2 the option <code>--hisat2</code> needs to be included as well.</p>

<h3 id="running-bismark-genome-preparation">Running <code>bismark_genome_preparation</code></h3>

<p><strong>USAGE:</strong> <code>bismark_genome_preparation [options] &lt;path_to_genome_folder&gt;</code></p>

<p>A typical command could look like this:</p>

<p><code>
bismark_genome_preparation --path_to_aligner /usr/bin/bowtie2/ --verbose /data/genomes/homo_sapiens/GRCh38/
</code></p>

<h2 id="ii-bismark-alignment-step">(II) Bismark Alignment Step</h2>

<p>This step represents the actual bisulfite alignment and methylation calling part. Bismark requires the user to specify only two things:</p>

<ol>
<li>The directory containing the genome of interest. This folder must contain the unmodified genome (as <code>.fa</code> or <code>.fasta</code> files) as well as the two bisulfite genome subdirectories which were generated in the Bismark Genome Preparations step (see above).
</li>
<li>The sequence file(s) to be analysed (in either <code>FastQ</code> or <code>FastA</code> format).
</li>
</ol>

<p>All other parameters are optional.</p>

<p>For each sequence file or each set of paired-end sequence files, Bismark produces one alignment and methylation call output file as well as a report file detailing alignment and methylation call statistics for your information and record keeping.</p>

<h3 id="running-bismark">Running <code>bismark</code></h3>

<p>Before running Bismark we recommend spending some time on quality control of the raw sequence files using <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. FastQC might be able to spot irregularities associated with your BS-Seq file, such as high base calling error rates or contaminating sequences such as PCR primers or Illumina adapters. Many sources of error impact detrimentally the alignment efficiencies and/or alignment positions, and thus possibly also affect the methylation calls and conclusions drawn from the experiment.</p>

<p>If no additional options are specified Bismark will use a set of default values, some of which are:</p>

<h3 id="using-bowtie">Using Bowtie 2:</h3>

<ul>
<li>Using Bowtie 2 is the default mode
</li>
<li>If no specific path to Bowtie 2 is specified it is assumed that the <code>bowtie2</code> executable is in the <code>PATH</code>
</li>
<li>Standard alignments use a multi-seed length of 20bp with 0 mismatches. These parameters can be modified using the options <code>-L</code> and <code>-N</code>, respectively
</li>
<li>Standard alignments use the default minimum alignment score function L,0,-0.2, i.e. f(x) = 0 + -0.2 * x (where x is the read length). For a read of 75bp this would mean that a read can have a lowest alignment score of -15 before an alignment would become invalid. This is roughly equal to 2 mismatches or ~2 indels of 1-2 bp in the read (or a combination thereof). The stringency can be set using the <code>--score_min &lt;func&gt;</code> function.
</li>
</ul>

<p>Even though the user is not required to specify additional alignment options it is often advisable to do so (e.g. when the default parameters are too strict). To see a full list of options please type <code>bismark --help</code> on the command line or see the Appendix at the end of this User Guide.</p>

<h3 id="directional-bs-seq-libraries-default">Directional BS-Seq libraries (default)</h3>

<p>Bisulfite treatment of DNA and subsequent PCR amplification can give rise to four (bisulfite converted) strands for a given locus. Depending on the adapters used, BS-Seq libraries can be constructed in two different ways:</p>

<ol>
<li>If a library is directional, only reads which are (bisulfite converted) versions of the original top strand (OT) or the original bottom strand (OB) will be sequenced. Even though the strands complementary to OT (CTOT) and OB (CTOB) are generated in the BS-PCR step they will not be sequenced as they carry the wrong kind of adapter at their 5’-end. By default, Bismark performs only 2 read alignments to the OT and OB strands, thereby ignoring alignments coming from the complementary strands as they should theoretically not be present in the BS-Seq library in question.
</li>
<li>Alternatively, BS-Seq libraries can be constructed so that all four different strands generated in the BS-PCR can and will end up in the sequencing library with roughly the same likelihood. In this case all four strands (OT, CTOT, OB, CTOB) can produce valid alignments and the library is called non- directional. Specifying <code>--non_directional</code> instructs Bismark to use all four alignment outputs.
</li>
</ol>

<p>To summarise again: alignments to the original top strand or to the strand complementary to the original top strand (OT and CTOT) will both yield methylation information for cytosines on the top strand. Alignments to the original bottom strand or to the strand complementary to the original bottom strand (OB and CTOB) will both yield methylation information for cytosines on the bottom strand, i.e. they will appear to yield methylation information for G positions on the top strand of the reference genome.</p>

<p>For more information about how to extract methylation information of the four different alignment strands please see below in the section on the Bismark methylation extractor.</p>

<p><strong>USAGE:</strong>
<code>
bismark [options] --genome &lt;genome_folder&gt; {-1 &lt;mates1&gt; -2 &lt;mates2&gt; | &lt;singles&gt;}
</code></p>

<p>A typical single-end analysis could look like this:</p>

<p><code>
bismark --genome /data/genomes/homo_sapiens/GRCh38/ sample.fastq.gz
</code></p>

<h3 id="what-does-the-bismark-output-look-like">What does the Bismark output look like?</h3>

<p>Since version 0.6.x the default output of Bismark is in BAM/SAM format (which is required to encode gapped alignments).</p>

<h3 id="bismark-bam-sam-output-default">Bismark BAM/SAM output (default)</h3>

<p>By default, Bismark generates SAM output for all alignment modes. Please note that reported quality values are encoded in Sanger format (Phred 33 scale), even if the input was in Phred64.</p>

<ol>
<li><code>QNAME</code> (seq-ID)
</li>
<li><code>FLAG</code> (this flag tries to take the strand a bisulfite read originated from into account (this is different from ordinary DNA alignment flags!))
</li>
<li><code>RNAME</code> (chromosome)
</li>
<li><code>POS</code> (start position)
</li>
<li><code>MAPQ</code> (calculated for Bowtie 2 and HISAT2)
</li>
<li><code>CIGAR</code>
</li>
<li><code>RNEXT</code>
</li>
<li><code>PNEXT</code>
</li>
<li><code>TLEN</code>
</li>
<li><code>SEQ</code>
</li>
<li><code>QUAL</code> (Phred33 scale)
</li>
<li><code>NM-tag</code> (edit distance to the reference)
</li>
<li><code>MD-tag</code> (base-by-base mismatches to the reference) (14) XM-tag (methylation call string)
</li>
<li><code>XR-tag</code> (read conversion state for the alignment) (16) XG-tag (genome conversion state for the alignment)
</li>
</ol>

<p>The mate read of paired-end alignments is written out as an additional separate line in the same format.</p>

<h3 id="data-visualisation">Data visualisation</h3>

<p>To see the location of the mapped reads the Bismark output file can be imported into a genome viewer, such as SeqMonk, using the chromosome, start and end positions (this can be useful to identify regions in the genome which display an artefactually high number of aligned reads). The alignment output can also be used to apply post-processing steps such as de-duplication (allowing only 1 read for each position in the genome to remove PCR artefacts) or filtering on the number of bisulfite conversion related non-bisulfite mismatches * (please note that such post-processing scripts are not part of the Bismark package).</p>

<blockquote>
  <p>* Bisulfite conversion related non-bisulfite mismatches are mismatch positions which have a C in the BS-read but a T in the genome; such mismatches may occur due to the way bisulfite read alignments are performed. Reads containing this kind of mismatches are not automatically removed from the alignment output in order not to introduce a bias for methylated reads. It should be noted that, even though no methylation calls are performed for these positions, reads containing bisulfite conversion related non-bisulfite mismatches might lead to false alignments if particularly lax alignment parameters were specified.</p>
</blockquote>

<h3 id="methylation-call">Methylation call</h3>

<p>The methylation call string contains a dot <code>.</code> for every position in the BS-read not involving a cytosine, or contains one of the following letters for the three different cytosine methylation contexts (UPPER CASE = METHYLATED, lower case = unmethylated):</p>

<ul>
<li><code>z</code> -  C in CpG context - unmethylated
</li>
<li><code>Z</code> -  C in CpG context - methylated
</li>
<li><code>x</code> -  C in CHG context - unmethylated
</li>
<li><code>X</code> -  C in CHG context - methylated
</li>
<li><code>h</code> - C in CHH context - unmethylated
</li>
<li><code>H</code> -  C in CHH context - methylated
</li>
<li><code>u</code> - C in Unknown context (CN or CHN) - unmethylated
</li>
<li><code>U</code> - C in Unknown context (CN or CHN) - methylated
</li>
<li><code>.</code> - not a C or irrelevant position
</li>
</ul>

<h2 id="iii-bismark-deduplication-step">(III) Bismark Deduplication Step</h2>

<p><code>USAGE: ./deduplicate_bismark [options] filename(s)</code></p>

<p>The script <code>deduplicate_bismark</code> is supposed to remove alignments to the same position in the genome from the Bismark mapping output (both single and paired-end SAM/BAM files), which can arise by e.g. excessive PCR amplification. Sequences which align to the same genomic position but on different strands are scored individually.</p>

<p><strong>It is important to note that deduplication is not recommended for RRBS, amplicon or other target enrichment-type libraries!</strong></p>

<p>In the default mode, the first alignment to a given position will be used irrespective of its methylation call. As the alignments are not ordered in any way this is near enough a random read for each position.</p>

<p>For <strong>single-end</strong> alignments, only the chromosome, start coordinate and strand of a read will be used for deduplication.</p>

<p>For <strong>paired-end</strong> alignments, the chromosome, the strand of a read pair, the start-coordinate of the first read as well as the start coordinate of the second read will be used for deduplication. This script expects the Bismark output to be in SAM/BAM format.</p>

<p><strong>Please note that for paired-end BAM files the deduplication script expects Read 1 and Read 2 to follow each other in consecutive lines!</strong> If the file has been sorted by position for whatever reason, please make sure that you resort it by read name first (e.g. using <code>samtools sort -n</code>)</p>

<h3 id="deduplication-using-umis-or-bars">Deduplication using UMIs or barcodes</h3>

<p>In addition to chromosome, start (and end position for paired-end libraries) position and strand orientation the option <code>--barcode</code> will also take a potential barcodes or UMIs (unique molecular identifiers) into consideration while deduplicating. The barcode needs to be the last element of the read ID and has to beseparated by a <code>:</code>, e.g.: </p>

<p><code>
MISEQ:14:000000000-A55D0:1:1101:18024:2858_1:N:0:CTCCT
</code></p>

<p>This option option is equivalent to using <a href="https://github.com/FelixKrueger/Umi-Grinder">UmiBam</a> in the following mode:
<code>UmiBam --umi input.bam</code>, however UmiBam has additional functionality such as a double UMI feature or the option to allow mismatches in the UMI(s).</p>

<h3 id="deduplication-of-multiple-files-of-the-same-library">Deduplication of multiple files of the same library</h3>

<p>When using the option <code>--multiple</code>, all specified input files are treated as a <strong>single</strong> sample and concatenated together for deduplication. This uses Unix <code>cat</code> for SAM files and <code>samtools cat</code> for BAM files. </p>

<p>Additional notes for BAM files: Although this works on either BAM or CRAM, all input files must be the same format as each other. The sequence dictionary of each input file must be identical, although this command does not check this. By default the header is taken from the first file to be concatenated.</p>

<h2 id="iv-bismark-methylation-extractor">(IV) Bismark methylation extractor</h2>

<p>Bismark comes with a supplementary <code>bismark_methylation_extractor</code> script which operates on Bismark result files and extracts the methylation call for every single C analysed. The position of every single C will be written out to a new output file, depending on its context (CpG, CHG or CHH), whereby methylated Cs will be labelled as forward reads (+), non-methylated Cs as reverse reads (-). The resulting files can be imported into a genome viewer such as SeqMonk (using the generic text import filter) and the analysis of methylation data can commence. Alternatively, the output of the methylation extractor can be transformed into a <code>bedGraph</code> and <code>coverage</code> file using the option <code>--bedGraph</code> (see also <code>--counts</code>). This step can also be accomplished from the methylation extractor output using the stand-alone script <code>bismark2bedGraph</code> (also part of the Bismark package available for download at <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">bioinformatics.babraham.ac.uk</a>). The <code>coverage</code> file can also be imported into SeqMonk directly using <code>Import Data &gt; Bismark (cov)</code>. Optionally, the bedGraph counts output can be used to generate a genome-wide cytosine report which reports the number on every single CpG (optionally every single cytosine) in the genome, irrespective of whether it was covered by any reads or not. As this type of report is informative for cytosines on both strands the output may be fairly large (~46mn CpG positions or >1.2bn total cytosine positions in the human genome...). The bedGraph to genome-wide cytosine report conversion can also be run individually using the stand- alone module <code>coverage2cytosine</code> (also part of the Bismark package available for download at <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">bioinformatics.babraham.ac.uk</a>).</p>

<p>As of Bismark version 0.6 or higher the default input format for the <code>bismark_methylation_extractor</code> is BAM/SAM (or potentially CRAM if you’ve got Samtools 1.2+ installed).</p>

<h3 id="the-methylation-extractor-output-looks-like-this-tab-separated">The methylation extractor output looks like this (tab separated):</h3>

<pre><code>1. seq-ID
2. methylation state
3. chromosome
4. start position (= end position)
5. methylation call
</code></pre>

<p>Methylated cytosines receive a <code>+</code> orientation, unmethylated cytosines receive a <code>-</code> orientation.</p>

<p>Examples for cytosines in CpG context:</p>

<pre><code>HWUSI-EAS611_0006:3:1:1058:15806#0/1 - 6 91793279 z
HWUSI-EAS611_0006:3:1:1058:17564#0/1 + 8 122855484 Z
</code></pre>

<p>Examples for cytosines in CHG context:</p>

<pre><code>HWUSI-EAS611_0006:3:1:1054:1405#0/1 - 7 89920171 x
HWUSI-EAS611_0006:3:1:1054:1405#0/1 + 7 89920172 X
</code></pre>

<p>Examples for cytosines in CHH context:</p>

<pre><code>HWUSI-EAS611_0006:3:1:1054:1405#0/1 - 7 89920184 h
</code></pre>

<p>The <code>bismark_methylation_extractor</code> comes with a few options, such as ignoring the first <int> number of positions in the methylation call string, e.g. to remove a restriction enzyme site (if RRBS is performed with non-directional BS-Seq libraries it might be required to remove reconstituted MspI sites at the beginning of each read as they will introduce a bias into the first methylation call). Another useful option for paired-end reads is called <code>--no_overlap</code> (on by default): specifying this option will extract the methylation calls of overlapping parts in the middle of paired-end reads only once (using the calls from the first read which is presumably the one with a lowest error rate).</p>

<p>For a full list of options type: <code>bismark_methylation_extractor --help</code> on the command line or refer to the Appendix section at the end of this User Guide.</p>

<h3 id="methylation-extractor-output">Methylation extractor output</h3>

<p>By default, the <code>bismark_methylation_extractor</code> discriminates between cytosines in CpG, CHG or CHH context. If desired, CHG and CHH contexts can be merged into a single non-CpG context by specifying the option <code>--merge_non_CpG</code> (as a word of warning, this might produce files with up to several hundred million lines...).</p>

<h3 id="strand-specific-methylation-output-files-default">Strand-specific methylation output files (default):</h3>

<p>As its default option, the <code>bismark_methylation_extractor</code> will produce a strand-specific output which will use the following abbreviations in the output file name to indicate the strand the alignment came from:</p>

<pre><code>OT    –  original top strand
CTOT  –  complementary to original top strand
OB    –  original bottom strand
CTOB  –  complementary to original bottom strand
</code></pre>

<p>Methylation calls from OT and CTOT will be informative for cytosine methylation positions on the original top strand, calls from OB and CTOB will be informative for cytosine methylation positions on the original bottom strand. Please note that specifying the <code>--directional</code> (the default mode) option in the Bismark alignment step will not report any alignments to the CTOT or CTOB strands.</p>

<p>As cytosines can exist in any of three different sequence contexts (CpG, CHG or CHH) the <code>bismark_methylation_extractor</code> default output will consist of 12 individual output files per input file (<code>CpG_OT_...</code>, <code>CpG_CTOT_...</code>, <code>CpG_OB_...</code> etc.).</p>

<h3 id="context-dependent-methylation-output-files-comprehensive-option">Context-dependent methylation output files (<code>--comprehensive</code> option):</h3>

<p>If strand-specific methylation is not of interest, all available methylation information can be pooled into a single context-dependent file (information from any of the four strands will be pooled). This will default to three output files (CpG-context, CHG-context and CHH-context), or result in 2 output files (CpG-context and Non-CpG-context) if <code>--merge_non_CpG</code> was selected (note that this can result in enormous file sizes for the non-CpG output).</p>

<p>Both strand-specific and context-dependent options can be combined with the <code>--merge_non_CpG</code> option.</p>

<h3 id="optional-bedgraph-output">Optional bedGraph output</h3>

<p>The Bismark methylation extractor can optionally also output a file in <a href="http://genome.ucsc.edu/goldenPath/help/bedgraph.html"><code>bedGraph</code></a> format which uses 0-based genomic start and 1- based end coordinates. The module <code>bismark2bedGraph</code> (part of the Bismark package) may also be run individually. It will be sorted by chromosomal coordinates and looks like this:
<code>
&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt;
</code></p>

<p>As the methylation percentage is <em>per se</em> not informative of the actual read coverage of detected methylated or unmethylated reads at a position, <code>bismark2bedGraph</code> also writes out a coverage file (using 1-based genomic genomic coordinates) that features two additional columns:
<code>
&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt; &lt;count methylated&gt; &lt;count unmethylated&gt;
</code></p>

<p>These two additional columns enable basically any downstream processing from the file. By default, this mode will only consider cytosines in CpG context, but it can be extended to cytosines in any sequence context by using the option <code>--CX</code> (cf. Appendix (III)).</p>

<h3 id="optional-genome-wide-cytosine-report-output">(Optional): genome-wide cytosine report output</h3>

<p>Starting from the <code>coverage</code> output, the Bismark methylation extractor can optionally also output a genome-wide cytosine methylation report. The module <code>coverage2cytosine</code> (part of the Bismark package) may also be run individually. It is also sorted by chromosomal coordinates but also contains the sequence context and is in the following format:
<code>
&lt;chromosome&gt; &lt;position&gt; &lt;strand&gt; &lt;count methylated&gt; &lt;count unmethylated&gt; &lt;C-context&gt; &lt;trinucleotide context&gt;
</code></p>

<p>The main difference to the <code>bedGraph</code> or <code>coverage</code> output is that <strong>every</strong> cytosine on both the top and bottom strands will be considered irrespective of whether they were actually covered by any reads in the experiment or not. For this to work one has to also specify the genome that was used for the <code>Bismark</code> alignments using the option <code>--genome_folder &lt;path&gt;</code>. As for the <code>bedGraph</code> mode, this will only consider cytosines in CpG context by default but can be extended to cytosines in any sequence context by using the option <code>--CX</code> (cf. Appendix (III)). Be aware though that this might mean an output with individual lines for more than 1.1 billion cytosines for any large mammalian genome...</p>

<h3 id="optional-nome-seq-or-scnmt-seq">(Optional): NOMe-seq or scNMT-seq</h3>

<p>The <code>coverage2cytosine</code> module can be instructed that a sample is a NOMe-seq (<strong>N</strong>ucleosome <strong>O</strong>ccupancy and <strong>Me</strong>thylome sequencing; https://genome.cshlp.org/content/22/12/2497.long) or scNMT-seq (<strong>s</strong>ingle-<strong>c</strong>ell <strong>N</strong>ucleosome, <strong>M</strong>ethylation and <strong>T</strong>ranscription sequencing (https://www.nature.com/articles/s41467-018-03149-4)) sample, where accessible DNA gets methylated in a GpC context (sets option <code>--gc</code> as well). The option <code>--nome-seq</code>:</p>

<p><code>
 (i) filters the genome-wide CpG-report to only output cytosines in ACG and TCG context
(ii) filters the GC context output to only report cytosines in GCA, GCC and GCT context
</code></p>

<p>Both of these measures aim to reduce unwanted biases, i.e. the influence of <code>G-CG</code> (intended) and <code>C-CG</code> (off-target) on endogenous CpG methylation, and the influence of CpG methylation on (the NOMe-seq specific) <code>GC</code> context methylation. <strong>PLEASE NOTE</strong> that NOMe-seq data requires a <code>.cov.gz</code> file as input which has been generated in non-CG mode!! (<code>--CX</code>), else the <code>GpC</code> output file will be empty...</p>

<h3 id="m-bias-plot">M-bias plot</h3>

<p>Starting with Bismark v0.8.0, the Bismark methylation extractor also produces a methylation bias plot which shows the methylation proportion across each possible position in the read (described in further detail in: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83">Hansen et al., Genome Biology, 2012, 13:R83</a>). The data for the M-bias plot is also written into a coverage text file (ending in <code>.cov</code> or <code>.cov.gz</code>) and is in the following format:</p>

<p><code>
&lt;read position&gt; &lt;count methylated&gt; &lt;count unmethylated&gt; &lt;% methylation&gt; &lt;total coverage&gt;
</code></p>

<p>This allows generating nice graphs by alternative means, e.g. using R or Excel. The plot is also drawn into a .png file which requires the Perl module GD::Graph (more specifically, both modules GD::Graph::lines and GD::Graph::colour are required); if GD::Graph cannot be found on the system, only the table will be printed. The plot also contains the absolute number of methylation calls (both methylated and unmethylated) per position. For paired-end reads two individual M-bias plots will be drawn.</p>

<p>The M-bias plot can for example show the methylation bias at the start of reads in PBAT-Seq experiments:</p>

<p><img src="Images/PBAT_SE_M-bias.jpg" alt="SE PBAT" /></p>

<p>For more on this topic please also see <a href="https://sequencing.qcfail.com/articles/mispriming-in-pbat-libraries-causes-methylation-bias-and-poor-mapping-efficiencies/">this post on QCFail.com</a>.</p>

<p>Or it can reveal a 3’-end-repair bias at the first couple of positions in read 2 of paired-end reads, like here:</p>

<p><img src="Images/PE_M-bias.jpg" alt="PE fill-in" /></p>

<p>For more on this topic please also see <a href="https://sequencing.qcfail.com/articles/library-end-repair-reaction-introduces-methylation-biases-in-paired-end-pe-bisulfite-seq-applications/">this post on QCFail.com</a>.</p>

<p>The M-bias plot should enable researchers to make an informed decision whether or not to leave the bias in the final data or to remove it (e.g. using the methylation extractor option <code>--ignore</code>).</p>

<h3 id="iii-running-bismark-methylation-extractor">(III) Running <code>bismark_methylation_extractor</code></h3>

<p><strong>USAGE:</strong> <code>bismark_methylation_extractor [options] &lt;filenames&gt;</code></p>

<p>A typical command for a single-end file could look like this: </p>

<p><code>
bismark_methylation_extractor --gzip sample_bismark_bt2.bam
</code></p>

<p>A typical command for a paired-end file could look like this:</p>

<p><code>
bismark_methylation_extractor --gzip sample_bismark_bt2_pe.bam
</code></p>

<p>The methylation extractor can also auto-detect the alignment mode and will set the options above automatically. A typical command including the optional <code>bedGraph</code> output could look like this:</p>

<p><code>
bismark_methylation_extractor --gzip --bedGraph --buffer_size 10G sample_bismark_bt2.bam
</code></p>

<p>A typical command including the optional genome-wide cytosine methylation report could look like this:</p>

<p><code>
bismark_methylation_extractor --gzip --bedGraph --buffer_size 10G --cytosine_report --genome_folder /path_to_genome_folder/ sample_bismark_bt2.bam
</code></p>

<h2 id="v-the-bismark-html-processing-report">(V) The Bismark HTML Processing Report</h2>

<p>The script <code>bismark2report</code> uses a Bismark alignment report, and optionally further reports of the Bismark suite such as deduplication, methylation extractor (splitting) or M-bias reports to generate a graphical HTML report page. If several Bismark reports are found in the same folder, a separate report will be generated for each of these, whereby the output filename is derived from the Bismark alignment report file. <code>bismark2report</code> attempts to find optional reports automatically based on the file basename. Here are an example <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/SE_report.html">single-end report</a> and <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/PE_report.html">paired-end report</a>.</p>

<p><strong>USAGE:</strong>
<code>
bismark2report [options]
</code></p>

<h4 id="options">OPTIONS</h4>

<ul>
<li><p><code>-o/--output &lt;filename&gt;</code></p>

<p>Name of the output file (optional). If not specified explicitly, the output filename will be derived                           from the Bismark alignment report file. Specifying an output filename only works if the HTML report is                            to be generated for a single Bismark alignment report (and potentially additional reports).</p></li>
<li><p><code>--dir &lt;directory&gt;</code></p></li>
</ul>

<p>Output directory. Output is written to the current directory if not specified explicitly.</p>

<ul>
<li><code>--alignment_report FILE</code>
</li>
</ul>

<p>If not specified explicitly, <code>bismark2report</code> attempts to find Bismark report file(s) in the current directory and produces a separate HTML report for each mapping report file. Based on the basename of the Bismark mapping report, <code>bismark2report</code> will also attempt to find the other Bismark reports (see below) for inclusion into the HTML report. Specifying a Bismark alignment report file is mandatory.</p>

<ul>
<li><code>--dedup_report FILE</code>
</li>
</ul>

<p>If not specified explicitly, <code>bismark2report</code> attempts to find a deduplication report file with the same                            basename as the Bismark mapping report (generated by <code>deduplicate_bismark</code>) in the current working directory. Including a deduplication report is optional, and using the FILE <em>'none'</em> will skip this step entirely.</p>

<ul>
<li><p><code>--splitting_report FILE</code></p>

<p>If not specified explicitly, <code>bismark2report</code> attempts to find a splitting report file with the same basename as the Bismark mapping report (generated by the Bismark methylation extractor) in the current working directory. Including a splitting report is optional, and using the FILE <em>'none'</em> will skip this step entirely.</p></li>
<li><p><code>--mbias_report FILE</code></p></li>
</ul>

<p>If not specified explicitly, <code>bismark2report</code> attempts to find a single M-bias report file with the same basename as the Bismark mapping report (generated by the Bismark methylation extractor) in the current working directory. Including an M-Bias report is optional, and using the FILE <em>'none'</em> will skip this step entirely.</p>

<ul>
<li><code>--nucleotide_report FILE</code>
</li>
</ul>

<p>If not specified explicitly, <code>bismark2report</code> attempts to find a single nucleotide coverage report file with the same basename as the Bismark mapping report (generated by Bismark with the option <code>--nucleotide_coverage</code>, or <code>bam2nuc</code> directly) in the current working directory. Including a nucleotide coverage statistics                            report is optional, and using the FILE <em>'none'</em> will skip this report entirely.</p>

<h2 id="vi-the-bismark-summary-report">(VI) The Bismark Summary Report</h2>

<p>This script uses Bismark report files of several (up to hundreds of!?) samples in a run folder to generate a graphical summary HTML report as well as a whopping big table (tab-delimited text) with all relevant alignment and methylation statistics which may be used for graphing purposes in R, Excel or the like. Unless certain BAM files are specified, <code>bismark2summary</code> first identifies Bismark BAM files in a folder (they need to use the Bismark naming conventions) and then automatically detects Bismark alignment, deduplication or methylation extractor (splitting) reports based on the input file basename. If splitting reports are found they overwrite the methylation statistics of the initial alignment report.</p>

<p><strong>USAGE:</strong>
<code>
bismark2summary [options]
</code></p>

<p>This command scans the current working directory for different Bismark alignment, deduplication and methylation extraction (splitting) reports to produce a graphical summary HTML report, as well as a data table, for all files in a directory. Here is a sample <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/bismark_summary_report.html">Bismark Summary Report</a>.</p>

<h4 id="arguments">ARGUMENTS:</h4>

<ul>
<li><p>BAM file(s)</p>

<p>Optional. If no BAM files are specified explicitly the current working directory is scanned for Bismark alignment files and their associated reports.</p></li>
</ul>

<h4 id="options">OPTIONS:</h4>

<ul>
<li><code>-o/--basename &lt;filename&gt;</code>
</li>
</ul>

<p>Basename of the output file (optional). Generate a text file with all relevant extracted values '<em>basename</em>.txt') as well as an HTML report ('<em>basename</em>.html'). If not specified explicitly, the basename is '<em>bismark_summary_report</em>'.</p>

<ul>
<li><code>--title &lt;string&gt;</code>
</li>
</ul>

<p>Optional HTML report title; use <code>--title "speech marks for text with spaces"</code>. Default: '<em>Bismark Summary Report</em>'.</p>

<ul>
<li><code>--version</code>
</li>
</ul>

<p>Displays version information and exits.</p>

<ul>
<li><code>--help</code> <br />
</li>
</ul>

<p>Displays this help message and exits.</p>

<h2 id="vii-bismark-nucleotide-coverage-report-bam-nuc">(VII) Bismark Nucleotide Coverage report (<code>bam2nuc</code>)</h2>

<p>The script <code>bam2nuc</code> reads BAM files and calculates the mono- and di-nucleotide coverage of the
reads (using the genomic sequence rather than the observed sequence in the reads themselves)
and compares it to the average genomic sequence composition. Reads harbouring InDels are not
taken into consideration. Mono- or dinucleotides containing Ns are ignored as well.</p>

<p><code>bam2nuc</code> handles both Bismark single-end and paired-end files (determined automatically). Both BAM and CRAM files should work as input, but please note that Samtools version 1.2 or higher is required for CRAM files.</p>

<p><strong>USAGE:</strong>
<code>
bam2nuc [options] --genome_folder &lt;path&gt; [input.(bam|cram)]
</code></p>

<ul>
<li><code>--dir</code>
</li>
</ul>

<p>Output directory. Output is written to the current directory if not specified explicitly.</p>

<ul>
<li><code>--genome_folder &lt;path&gt;</code>
</li>
</ul>

<p>Enter the genome folder you wish to use to extract sequences from (full path only). Accepted formats are FastA files ending with '.fa' or '.fasta'. Specifying a genome folder path is mandatory.</p>

<ul>
<li><code>--samtools_path</code>
</li>
</ul>

<p>The path to your Samtools installation, e.g. /home/user/samtools/. Does not need to be specified explicitly if Samtools is in the <code>PATH</code> already</p>

<ul>
<li><code>--genomic_composition_only</code>
</li>
</ul>

<p>Only calculate and extract the genomic sequence composition and exit thereafter. This option will attempt to write the genomic composition table 'genomic_nucleotide_frequencies.txt' to the genome folder or to the output directory instead if that doesn't succeed.</p>

<ul>
<li><code>--help</code>
</li>
</ul>

<p>Displays this help message and exits</p>

<p><strong>GENOMIC COMPOSITION</strong></p>

<p>Since the calculation of the average genomic (di-)nucleotide composition may take a while, <code>bam2nuc</code> attempts to write out a file called 'genomic_nucleotide_frequencies.txt' to the genome folder if it wasn't there already. The next time <code>bam2nuc</code> is run it will then use this file instead of calculating the average genome composition again. If writing to the genome folder fails (e.g. because of permission issues) it will be written out to the output directory instead.</p>

<p><strong>OUTPUT FORMAT</strong></p>

<p><code>bam2nuc</code> writes out a file ending in <em>.nucleotide_stats.txt</em> in the following format (tab-delimited):</p>

<pre><code>(di-)nucleotide count sample    percent sample  count genomic   percent genomic coverage
A       14541   30.91   3768086 30.98   0.004
C       8893    18.90   2321832 19.09   0.004
G       9019    19.17   2318192 19.06   0.004
T       14597   31.02   3754886 30.87   0.004
AA      5008    10.86   1321485 10.86   0.004
AC      2355    5.11    639783  5.26    0.004
AG      2692    5.84    709163  5.83    0.004
AT      4191    9.09    1097652 9.02    0.004
CA      2912    6.32    786744  6.47    0.004
CC      1812    3.93    473900  3.90    0.004
CG      1341    2.91    355535  2.92    0.004
CT      2659    5.77    705653  5.80    0.004
GA      2903    6.30    756411  6.22    0.004
GC      1724    3.74    453607  3.73    0.004
GG      1817    3.94    470732  3.87    0.004
GT      2402    5.21    637436  5.24    0.004
TA      3419    7.42    903441  7.43    0.004
TC      2823    6.12    754531  6.20    0.004
TG      2996    6.50    782761  6.44    0.004
TT      5055    10.96   1314144 10.80   0.004
</code></pre>

<p>This file is picked up and plotted by <code>bismark2report</code> automatically if found in the folder in the following manner:</p>

<p><img src="Images/nucleocoverage.png" alt="Nucleotide Coverage Plot" /></p>

<h2 id="viii-filtering-out-non-bisulfite-converted-reads-filter-non-conversion">(VIII) Filtering out non-bisulfite converted reads (<code>filter_non_conversion</code>)</h2>

<p>Filtering incomplete bisulfite conversion from Bismark BAM files (optional). This script examines the methylation calls of reads, or read pairs for paired-end sequencing, and filters out reads that exceed a certain threshold of methylated calls in non-CG context (the default is 3). By default, <code>filter_non_conversion</code> looks for a certain number of methylated non-CG calls, but a percentage methylation cutoff may be specified alternatively.</p>

<p><strong>Please Note</strong>: Be aware that this kind of filtering is not advisable - and <em>will</em> introduce biases - if you work with organisms which exhibit any appreciable levels of non-CG methylation (e.g. most plants).</p>

<p>Writes out a file called <em>nonCG_filtered.bam</em>, also a file called <em>nonCG_removed_seqs.bam</em> as well as a short report how many sequences have been analysed and removed.</p>

<p><strong>USAGE:</strong></p>

<p><code>
filter_non_conversion [options] [Bismark BAM files]
</code></p>

<p><strong>Please also note</strong> that for paired-end BAM files <code>filter_non_conversion</code> expects Read 1 and Read 2 to follow each other in consecutive lines! If the file has been sorted by position make sure that you resort it by read name first (e.g. using <code>samtools sort -n</code>)</p>

<ul>
<li><code>-s/--single</code>
</li>
</ul>

<p>Deduplicate single-end Bismark BAM files. If not specified the library type is auto-detected.</p>

<ul>
<li><code>-p/--paired</code>
</li>
</ul>

<p>Deduplicate paired-end Bismark BAM files. If not specified the library type is auto-detected.</p>

<ul>
<li><code>--threshold [int]</code>
</li>
</ul>

<p>The number of methylated cytosines in non-CG context at which reads or read pairs are filtered out. For paired-end files either Read 1 or Read 2 can fail the entire read pair. [Default: 3].</p>

<ul>
<li><code>--percentage_cutoff [int]</code>
</li>
</ul>

<p>Instead of filtering on an absolute count of methylated cytosines in non-CG context (see <code>--threshold [int]</code>) this option allows you to define an overall percentage of methylation in non-CG context (both CHH and CHG) which, if reached or exceeded, results in the read or read pair being filtered out. For paired-end files either Read 1 or Read 2 can fail the entire read pair. Also requires a minimum number of cytosines in non-CG context to make confident filtering choices (see <code>--minimum_count [int]</code>).</p>

<ul>
<li><code>--minimum_count [int]</code>
</li>
</ul>

<p>At least this number of cytosines in non-CG context (CHH or CHG) have to be seen in a read (irrespective of their methylation state) before the <code>--percentage_cutoff</code> filter kicks in. [Default: 5].</p>

<ul>
<li><code>--consecutive</code>
</li>
</ul>

<p>Non-CG methylation has to be found on consecutive non-CGs. Any kind of unmethylated cytosine (in any context) resets the methylated non-CG counter to 0. [Default: OFF].</p>

<ul>
<li><code>--samtools_path</code>
</li>
</ul>

<p>The path to your Samtools installation, e.g. /home/user/samtools/. Does not need to be specified explicitly if Samtools is in the PATH already.</p>

<ul>
<li><code>--help</code>
</li>
</ul>

<p>Displays this help text end exits.</p>

<ul>
<li><code>--version</code>
</li>
</ul>

<p>Displays version information and exits.</p>

<p>If you get stuck at any point or have any questions or comments please contact me via e-mail: <a href="felix.krueger@babraham.ac.uk">felix.krueger@babraham.ac.uk</a></p>

<h2 id="ix-notes-about-different-library-types-and-commercial-kits">(IX) Notes about different library types and commercial kits</h2>

<p>Here is a table summarising general recommendations for different library types and/or different commercially available kits. Some more specific notes can be found below.</p>

<table>
    <thead>
        <tr>
            <th align="left">Technique</th>
            <th align="center">5' Trimming</th>
            <th align="center">3' Trimming</th>
            <th align="center">Mapping</th>
            <th align="center">Deduplication</th>
            <th align="center">Extraction</th>
    </tr>
    </thead>
    <tbody>
        <tr>
            <td align="left">BS-Seq</td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><code>--ignore_r2 2</code></td>
        </tr>
        <tr>
            <td align="left">RRBS</td>
            <td align="center"><code>--rrbs</code> (R2 only)</td>
            <td align="center"><code>--rrbs</code> (R1 only)</td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="x" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" ios-version="6.0">❌</g-emoji></td>
            <td align="center">(<code>--ignore_r2 2</code>)</td>
        </tr>
        <tr>
            <td align="left">RRBS (NuGEN Ovation)</td>
            <td align="center">special processing</td>
            <td align="center">special processing</td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="x" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" ios-version="6.0">❌</g-emoji></td>
            <td align="center"><code>--ignore_r2 2</code></td>
        </tr>
        <tr>
            <td align="left">PBAT</td>
            <td align="center">6N / 9N</td>
            <td align="center">(6N / 9N)</td>
            <td align="center"><code>--pbat</code></td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
        </tr>
        <tr>
            <td align="left">single-cell (scBS-Seq)</td>
            <td align="center">6N</td>
            <td align="center">(6N)</td>
            <td align="center"><code>--non_directional</code>; single-end mode</td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
        </tr>
        <tr>
            <td align="left">TruSeq (EpiGnome)</td>
            <td align="center">8 bp</td>
            <td align="center">(8 bp)</td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
        </tr>
        <tr>
            <td align="left">Accel-NGS (Swift)</td>
            <td align="center">10 bp</td>
            <td align="center">(10 bp)</td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
        </tr>
        <tr>
            <td align="left">Zymo    Pico-Methyl</td>
            <td align="center">10 bp</td>
            <td align="center">(10 bp)</td>
            <td align="center"><code>--non_directional</code></td>
            <td align="center"><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji></td>
            <td align="center"><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji></td>
        </tr>
    </tbody>
</table>

<ul>
<li><g-emoji alias="white_large_square" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2b1c.png" ios-version="6.0">⬜️</g-emoji> -
Default settings (nothing in particular is required, just use Trim Galore or Bismark default parameters)
</li>
<li><g-emoji alias="white_check_mark" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/2705.png" ios-version="6.0">✅</g-emoji> -
Yes, please!
</li>
<li><g-emoji alias="x" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" ios-version="6.0">❌</g-emoji> -
No, absolutely not!
</li>
</ul>

<p><strong>5' Trimming</strong> can be accomplished with Trim Galore using: </p>

<p><code>--clip_r1 &lt;NUMBER&gt;</code> (Read 1) or </p>

<p><code>--clip_r2 &lt;NUMBER&gt;</code> (Read 2)</p>

<p><strong>3' Trimming</strong> can be accomplished with Trim Galore using:</p>

<p><code>--three_prime_clip_r1 &lt;NUMBER&gt;</code> (Read 1) or </p>

<p><code>--three_prime_clip_r2 &lt;NUMBER&gt;</code> (Read 2).</p>

<h4 id="specific-library-kit-notes">SPECIFIC LIBRARY/KIT NOTES</h4>

<h4 id="rrbs">RRBS</h4>

<p>RRBS is a specialised technique to only look at CpG rich regions of the genome by using the restriction enzyme MspI (please see this <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/RRBS_Guide.pdf">RRBS Guide</a> for some more specifics regarding data processing). For reasons explained in the RRBS-guide, the second last position of all reads before reading into the Illumina adapter exhibits an artificially (not methylated) methylation state as a result of the end-repair reaction. The option <code>--rrbs</code> within Trim Galore removes 2 extra bases whenever adapter contamination has been detected. This 3' end trimming that needs to be carried out for single-end runs or Read 1 of paired-end libraries. Read 2 of paired-end libraries is however not affected by this 3' bias, but instead the first couple of positions on the 5' end of Read 2 suffer from the read-through problem as Read 1 (Read 2 is a mere copy of Read 1), so Read 2 needs to have the <strong>first</strong> 2 bp removed instead. As of the current development version of <a href="https://github.com/FelixKrueger/TrimGalore">Trim Galore</a> (v0.4.2_dev; 12/16/2016) the option <code>--rrbs</code> removes:
- 2 bp from the 3' end of single-end and Read 1 of paired-end reads in addition to adapter contamination, and
- 2 bp from the 5' end of Read 2 of paired-end reads</p>

<h4 id="rrbs-nugen-ovation-methyl-seq-system">RRBS NuGEN Ovation Methyl-Seq System</h4>

<p>(<a href="http://www.nugen.com/products/ovation-rrbs-methyl-seq-system">Manufacturer's page</a>)</p>

<p>Owing to the fact that the NuGEN Ovation kit attaches a varying number of nucleotides (0-3) after each MspI site Trim Galore should be run <em>WITHOUT</em> the option <code>--rrbs</code>. The trimming is accomplished in a subsequent diversity trimming step afterwards, please see the manufacturer's manual for more details.</p>

<h4 id="pbat">PBAT</h4>

<p>The amount of bases that need to be trimmed from the 5' end depends on the length of the oligo used for random priming, which - as we know - isn't all that random, and in fact causes <a href="https://sequencing.qcfail.com/articles/mispriming-in-pbat-libraries-causes-methylation-bias-and-poor-mapping-efficiencies/">misalignments and methylation biases</a>. While the original PBAT paper used 4N oligoes, these days 6N or 9N seem to be most common. Please also see the section <em>3' Trimming in general</em> below.</p>

<h4 id="single-cell">Single-cell</h4>

<p>The <a href="http://www.nature.com/nmeth/journal/v11/n8/full/nmeth.3035.html">scBS-Seq method</a> uses a PBAT-type protocol but employs five rounds of sequence capture and elongation to amplify the starting material so all four different bisulfite strands (OT, CTOT, OB, CTOB) are sequenced. Since 6N oligos are used to for the random priming step, 6 bp need to be removed from the 5' ends. Since scBS and PBAT libraries tend to result in <a href="https://sequencing.qcfail.com/articles/pbat-libraries-may-generate-chimaeric-read-pairs/">chimaeric fragments</a> we tend to treat scBS-Seq as single-end reads always. Please also see the section <em>3' Trimming in general</em> below.</p>

<h4 id="truseq-dna-methylation-kit-formerly-epignome">TruSeq DNA-Methylation Kit (formerly EpiGnome)</h4>

<p>(<a href="http://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/truseq-dna-methylation.html">Manufacturer's page</a>)
This Illumina kit (previously known as EpiGnome kit from epicentre) also employs a post-bisulfite strategy using 6N oligos, but in contrast to the PBAT technique only the standard original top and bottom strands (OT and OB) are sequenced, meaning that Bismark can be run in default (= directional) mode. Even though the random priming is performed with 6N oligoes we often saw that the methylation bias extends to 7 or 8 bp, so trimming 8 bp off the 5' end(s) is recommended initially. Please do have a look at the M-bias plots nevertheless to see of more bases need removing/ignoring during the methylation extraction process. Please also see the section <em>3' Trimming in general</em> below.</p>

<h4 id="zymo-pico-methyl-seq">Zymo Pico Methyl-Seq</h4>

<p>(<a href="https://www.zymoresearch.com/epigenetics/dna-methylation/genome-wide-5-mc-analysis/pico-methyl-seq-library-prep-kit">Manufacturer's page</a>)
The Pico Methyl-Seq kit also uses a random priming step similar to the PBAT // single-cell methods above. This kit uses random tetramers (4N) for the amplification step, however the biases seen in the base composition and M-bias plots indicate that one should trim off at least the first 10 bp from each read. This kit performs three rounds of amplification which yields non-directional libraries (similar to the scBS-Seq protocol), so all four different bisulfite strands (OT, CTOT, OB, CTOB) are present in the library. According to the manufacturer, the library construction is designed for a starting input material of 100 ng, but can be scaled up or down (to 100 pg). Please also see the section <em>3' Trimming in general</em> below.</p>

<h4 id="swift">Swift</h4>

<p><a href="https://swiftbiosci.com/products/accel-ngs-methyl-seq-dna-library-kit/">Manufacturer's page</a>
The Accel-NGS Methyl-Seq protocol uses Adaptase technology for capturing single-stranded DNA in an unbiased (again, not that unbiased actually...) manner. Also here, the first ~10 bp show extreme biases in sequence composition and M-bias, so trimming off at <em>least</em> 10 bp is advisable (please check the M-bias plot if even more is needed). Please also see the section <em>3' Trimming in general</em> below.</p>

<h4 id="random-priming-and-trimming-in-general">Random priming and 3' Trimming in general</h4>

<p>As we have seen before, the random priming of post-bisulfite methods (such as PBAT, scBS-Seq, EpiGnome, Pico Methyl, Accel etc.) introduces <a href="https://sequencing.qcfail.com/articles/mispriming-in-pbat-libraries-causes-methylation-bias-and-poor-mapping-efficiencies/">errors, indels and methylation biases</a> that may detrimentally affect your mapping efficiencies and methylation calls. These problems are fairly easy to spot at the 5' ends of reads because all reads will equally suffer from the problems at the same positions at the start (5' end) of reads. 
The same problems of random priming (indels, mispriming) will however most likely occur on both sides of the fragment to be sequenced, but it is doubtful that one would be able to spot these problems on the 3' end of reads because the problems would be expected on the 3' end of reads just before reading through into the adapter, and this may occur 
 - at different positions in the read (depending on how short the fragment was)
 - at different positions within the read because of quality trimming in addition to adapter read-through contamination 
 - not at all within the read length (whenever a fragment is longer than the sequenced read length)
 - at the 3' end even without hitting the adapter (i.e. just before the adapter)</p>

<p>I guess there is a trade-off between accepting that a certain proportion of the reads may have a few biased biased positions towards their 5' ends, and preemptively trimming the 3' end by the same amount of bases as the 5' end. As a general rule it is probably safe to say that the shorter the average insert size of a library - the more of a problem the bias is. We have e.g. seen Pico Methyl libraries where ~80% of all fragments were shorter than 100bp, so a 2x125bp run would most likely be affected by the random priming bias on the 5' and 3' ends in nearly all fragments sequenced. We realise that trimming off say 10 bp from the 5' end and 3' end of a 100 bp read already removes 20% of the actually sequenced data, but this is the price you have to pay for using post-bisulfite kits...</p>

<p>For these reasons we have put the <em>3' Trimming</em> values in the table above in <em>(parentheses)</em> as a reminder that you <strong>should</strong> probably perform 3' trimming of the data as well. </p>

<h1 id="appendix-full-list-of-options">4) APPENDIX - Full list of options</h1>

<h2 id="appendix-i-bismark-genome-preparation">Appendix (I): Bismark Genome Preparation</h2>

<p>A full list of options can also be viewed by typing: <code>bismark_genome_preparation --help</code></p>

<h4 id="usage-bismark-genome-preparation-options-lt-arguments-gt">USAGE: <code>bismark_genome_preparation [options] &lt;arguments&gt;</code></h4>

<h4 id="options">OPTIONS:</h4>

<ul>
<li><p><code>--help</code></p>

<p>Displays help text.</p></li>
<li><p><code>--version</code></p>

<p>Displays version information and exits.</p></li>
<li><p><code>--verbose</code></p>

<p>Print verbose output for more details or debugging.</p></li>
<li><p><code>--path_to_aligner &lt;/../../&gt;</code></p>

<p>The full path to the Bowtie 2 or HISAT2 installation folder on your system (depending on which aligner/indexer you intend to use; please note that thi is the folder and not any executable). Unless this path is specified, it is assumed that the aligner in question (Bowtie 2/HISAT2) is in the PATH.</p></li>
<li><p><code>--bowtie2</code></p>

<p>This will create bisulfite indexes for use with Bowtie 2. Recommended for most bisulfite sequencing applications (Default: ON).</p></li>
<li><p><code>--hisat2</code></p>

<p>This will create bisulfite indexes for use with HISAT2. At the time of writing, this is still largely unchartered territory, and only recommended for specialist applications such as RNA-methylation analyses or SLAM-seq type applications (see also: --slam). (Default: OFF).</p></li>
<li><p><code>--single_fasta</code></p>

<p>Instruct the Bismark Indexer to write the converted genomes into single-entry FastA files instead of making one multi-FastA file (MFA) per chromosome. This might be useful if individual bisulfite converted chromosomes are needed (e.g. for debugging), however it can cause a problem with indexing if the number of chromosomes is vast (this is likely to be in the range of several thousand files; operating systems can only handle lists up to a certain length. Some newly assembled genomes may contain 20000-500000 contig of scaffold files which do exceed this list length limit).</p></li>
<li><p><code>--genomic_composition</code></p>

<p>Calculate and extract the genomic sequence composition for mono- and di-nucleotides and write the genomic composition table <em>genomic_nucleotide_frequencies.txt</em> to the genome folder. This may be useful later on when using <code>bam2nuc</code> or the Bismark option <code>--nucleotide_coverage</code>.</p></li>
<li><p><code>--slam</code></p>

<p>Instead of performing an in-silico bisulfite conversion, this mode transforms T to C (forward strand), or A to G (reverse strand). The folder structure and rest of the indexing process is currently exactly the same as for bisulfite sequences, but this might change at some point. This means that a genome prepared in --slam mode is currently indistinguishable from a true Bisulfite Genome, so please make sure you name the genome folder appropriately to avoid confusion.</p></li>
</ul>

<h4 id="arguments">ARGUMENTS:</h4>

<ul>
<li><p><code>&lt;path_to_genome_folder&gt;</code></p>

<p>The path to the folder containing the genome to be bisulfite converted (this may be an absolute or relative path). Bismark Genome Preparation expects one or more <code>FastA</code> files in the folder (valid file extensions: <code>.fa</code> or <code>.fasta</code>).</p></li>
</ul>

<h2 id="appendix-ii-bismark">Appendix (II): Bismark</h2>

<p>A brief description of Bismark and a full list of options can also be viewed by typing:
<code>bismark --help</code></p>

<h4 id="usage">USAGE:</h4>

<p><code>
bismark [options] --genome &lt;genome_folder&gt; {-1 &lt;mates1&gt; -2 &lt;mates2&gt; | &lt;singles&gt;}
</code></p>

<h4 id="arguments">ARGUMENTS:</h4>

<ul>
<li><p><code>&lt;genome_folder&gt;</code></p>

<p>The full path to the folder containing the unmodified reference genome as well as the sub folders created by the <code>bismark_genome_preparation</code> script (<code>Bisulfite_Genome/CT_conversion/</code> and <code>Bisulfite_Genome/GA_conversion/</code>). Bismark expects one or more <code>FastA</code> files in this folder (file extension: <code>.fa</code> or <code>.fasta</code>). The path to the genome folder can be relative or absolute. The path may also be set as <code>--genome_folder /path/to/genome/folder/</code>.</p></li>
<li><p><code>-1 &lt;mates1&gt;</code></p>

<p>Comma-separated list of files containing the #1 mates (filename usually includes <code>_1</code>), e.g. <code>flyA_1.fq</code>, <code>flyB_1.fq</code>). Sequences specified with this option must correspond file-for-file and read-for-read with those specified in <code>&lt;mates2&gt;</code>. Reads may be a mix of different lengths. Bismark will produce one mapping result and one report file per paired-end input file pair.</p></li>
<li><p><code>-2 &lt;mates2&gt;</code></p>

<p>Comma-separated list of files containing the #2 mates (filename usually includes <code>_2</code>), e.g. <code>flyA_2.fq</code>, <code>flyB_2.fq</code>). Sequences specified with this option must correspond file-for-file and read-for-read with those specified in <code>&lt;mates1&gt;</code>. Reads may be a mix of different lengths.</p></li>
<li><p><code>&lt;singles&gt;</code></p>

<p>A comma or space separated list of files containing the reads to be aligned (e.g. <code>lane1.fq</code>,<code>lane2.fq</code> <code>lane3.fq</code>). Reads may be a mix of different lengths. Bismark will produce one mapping result and one report file per input file.</p></li>
</ul>

<h4 id="options">OPTIONS:</h4>

<h5 id="input">Input:</h5>

<ul>
<li><p><code>--se &lt;list&gt;</code> / <code>--single_end &lt;list&gt;</code></p>

<p>Sets single-end mapping mode explicitly giving a list of file names as <list>. The filenames may be provided as a comma [<code>,</code>] or colon [<code>:</code>]-separated list.</p></li>
<li><p><code>-q</code> / <code>--fastq</code></p>

<p>The query input files (specified as <code>&lt;mate1&gt;</code>, <code>&lt;mate2&gt;</code> or <code>&lt;singles&gt;</code> are FastQ files (usually having extension <code>.fq</code> or <code>.fastq</code>). Input files may also be <code>gzip</code> compressed (ending in <code>.gz</code>). This is the default. See also <code>--solexa-quals</code> and <code>--integer-quals</code>.</p></li>
<li><p><code>-f</code> / <code>--fasta</code></p>

<p>The query input files (specified as <code>&lt;mate1&gt;</code>, <code>&lt;mate2&gt;</code> or <code>&lt;singles&gt;</code> are FastA files (usually having extension <code>.fa</code>, <code>.mfa</code>, <code>.fna</code> or similar). Input files may also be <code>gzip</code> compressed (ending in <code>.gz</code>).  All quality values are assumed to be 40 on the Phred scale. FASTA files are expected to contain the read name and the sequence on a single line each (and not spread over several lines)</p></li>
<li><p><code>-s &lt;int&gt;</code> / <code>--skip &lt;int&gt;</code></p>

<p>Skip (i.e. do not align) the first &lt;int> reads or read pairs from the input.</p></li>
<li><p><code>-u &lt;int&gt;</code> / <code>--upto &lt;int&gt;</code></p>

<p>Only aligns the first &lt;int> reads or read pairs from the input. Default: no limit.</p></li>
<li><p><code>--phred33-quals</code></p>

<p>FastQ qualities are ASCII chars equal to the Phred quality plus 33. Default: ON.</p></li>
<li><p><code>--phred64-quals</code></p>

<p>FastQ qualities are ASCII chars equal to the Phred quality plus 64. Default: OFF.</p></li>
<li><p><code>--path_to_bowtie2</code></p>

<p>The full path <code>&lt;/../../&gt;</code> to the Bowtie 2 installation on your system. If not specified it will be assumed that Bowtie 2 is in the <code>PATH</code>.</p>

<ul>
<li><code>--path_to_hisat2</code>
</li>
</ul>

<p>The full path <code>&lt;/../../&gt;</code> to the HISAT2 installation on your system. If not specified it will be assumed that HISAT2 is in the <code>PATH</code>.</p></li>
</ul>

<h5 id="alignment">Alignment:</h5>

<ul>
<li><p><code>-N &lt;int&gt;</code></p>

<p>Sets the number of mismatches to be allowed in a seed alignment during multiseed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often <em>much</em> slower) but increases sensitivity. Default: 0.</p></li>
<li><p><code>-L &lt;int&gt;</code></p>

<p>Sets the length of the seed substrings to align during multiseed alignment. Smaller values make alignment slower but more sensitive. Default: the <code>--sensitive</code> preset of Bowtie 2 is used by default, which sets <code>-L</code> to 20.</p></li>
<li><p><code>--ignore-quals</code></p>

<p>When calculating a mismatch penalty, always consider the quality value at the mismatched position to be the highest possible, regardless of the actual value. i.e. input is  treated as though all quality values are high. This is also the default behaviour when the input doesn't specify quality values (e.g. in <code>-f</code> mode). For bisulfite alignments in Bismark, this option is invariably turned on by default.</p></li>
<li><p><code>-I &lt;int&gt;</code> / <code>--minins &lt;int&gt;</code></p>

<p>The minimum insert size for valid paired-end alignments. E.g. if <code>-I 60</code> is specified and a paired-end alignment consists of two 20-bp alignments in the appropriate orientation with a 20-bp gap between them, that alignment is considered valid (as long as <code>-X</code> is also satisfied). A 19-bp gap would not be valid in that case. Default: 0.</p></li>
<li><p><code>-X &lt;int&gt;</code> / <code>--maxins &lt;int&gt;</code></p>

<p>The maximum insert size for valid paired-end alignments. E.g. if <code>-X 100</code> is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, that alignment is considered valid (as long as <code>-I</code> is also satisfied). A 61-bp gap would not be valid in that case. Default: 500.</p></li>
<li><p><code>--multicore &lt;int&gt;</code></p>

<p>Sets the number of parallel instances of Bismark to be run concurrently. This forks the Bismark alignment step very early on so that each individual Spawn of Bismark processes only every n-th sequence (n being set by <code>--multicore</code>). Once all processes have completed, the individual BAM files, mapping reports, unmapped or ambiguous FastQ files are merged into single files in very much the same way as they would have been generated running Bismark conventionally with only a single instance.</p>

<p>If system resources are plentiful this is a viable option to speed up the alignment process (we observed a near linear speed increase for up to <code>--multicore 8</code> tested). However, please note that a typical Bismark run will use several cores already (Bismark itself, 2 or 4 threads for Bowtie/Bowtie2, Samtools, gzip etc...) and ~10-16GB of memory per thread depending on the choice of aligner and genome.
<strong>WARNING:</strong> Bismark Parallel is <strong>resource hungry</strong>! Each value of <code>--multicore</code> specified will effectively lead to a linear increase in compute and memory requirements, so <code>--multicore 4</code> for e.g. the GRCm38 mouse genome will probably use ~20 cores and eat ~48GB of RAM, but at the same time reduce the alignment time to ~25-30%. <em>You have been warned</em>.</p></li>
</ul>

<h5 id="output">Output:</h5>

<ul>
<li><p><code>--non_directional</code></p>

<p>The sequencing library was constructed in a non strand-specific manner, alignments to all four bisulfite strands will be reported.
(The current Illumina protocol for BS-Seq is directional, in which case the strands complementary to the original strands are merely theoretical and should not exist in reality. Specifying directional alignments (which is the default) will only run 2 alignment threads to the original top (OT) or bottom (OB) strands in parallel and report these alignments. This is the recommended option for strand-specific libraries).
Default: OFF</p></li>
<li><p><code>--pbat</code></p>

<p>This option may be used for PBAT-Seq libraries (Post-Bisulfite Adapter Tagging; Kobayashi et al., PLoS Genetics, 2012). This is essentially the exact opposite of alignments in 'directional' mode, as it will only launch two alignment threads to the CTOT and CTOB strands instead of the normal OT and OB ones. Use this option only if you are certain that your libraries were constructed following a PBAT protocol (if you don't know what PBAT-Seq is you should not specify this option). The option <code>--pbat</code> works only for FastQ files and uncompressed temporary files.</p></li>
<li><p><code>--sam-no-hd</code></p>

<p>Suppress SAM header lines (starting with @). This might be useful when very large input files are split up into several smaller files to run concurrently and the output files are to be merged afterwards.</p></li>
<li><p><code>--rg_tag</code></p>

<p>Write out a Read Group tag to the resulting SAM/BAM file. This will write the following line to the SAM header:</p>

<p>@RG PL: ILLUMINA ID:SAMPLE SM:SAMPLE</p>

<p>to set ID and SM see <code>--rg_id</code> and <code>--rg_sample</code>. In addition each read receives an <code>RG:Z:RG-ID</code> tag.  Default: OFF (to not inflate file sizes).</p></li>
<li><p><code>--rg_id &lt;string&gt;</code></p>

<p>Sets the ID field in the <code>@RG</code> header line. Default: SAMPLE</p></li>
<li><p><code>--rg_sample &lt;string&gt;</code></p>

<p>Sets the SM field in the <code>@RG</code> header line; can't be set without setting <code>--rg_id</code> as well. Default: SAMPLE</p></li>
<li><p><code>--quiet</code></p>

<p>Print nothing besides alignments.</p></li>
<li><p><code>--un</code></p>

<p>Write all reads that could not be aligned to the file <code>_unmapped_reads.fq.gz</code> in the output directory. Written reads will appear as they did in the input, without any translation of quality values that may have taken place within <code>Bowtie</code> or <code>Bismark</code>. Paired-end reads will be written to two parallel files with <code>_1</code> and <code>_2</code> inserted in their filenames, i.e. <code>unmapped_reads_1.fq.gz</code> and <code>unmapped_reads_2.fq.gz</code>. Reads with more than one valid alignment with the same number of lowest mismatches (ambiguous mapping) are also written to <code>unmapped_reads.fq.gz</code> unless <code>--ambiguous</code> is also specified.</p></li>
<li><p><code>--ambiguous</code></p>

<p>Write all reads which produce more than one valid alignment with the same number of lowest mismatches or other reads that fail to align uniquely to <code>_ambiguous_reads.fq</code>. Written reads will appear as they did in the input, without any of the translation of quality values that may have taken place within <code>Bowtie</code> or <code>Bismark</code>. Paired-end reads will be written to two parallel files with <code>_1</code> and <code>_2</code> inserted in their filenames, i.e. <code>_ambiguous_reads_1.fq</code> and <code>_ambiguous_reads_2.fq</code>. These reads are not written to the file specified with <code>--un</code>.</p></li>
<li><p><code>-o/--output_dir &lt;dir&gt;</code></p>

<p>Write all output files into this directory. By default the output files will be written into the same folder as the input file. If the specified folder does not exist, Bismark will attempt to create it first. The path to the output folder can be either relative or absolute.</p></li>
<li><p><code>--temp_dir &lt;dir&gt;</code></p>

<p>Write temporary files to this directory instead of into the same directory as the input files. If the specified folder does not exist, Bismark will attempt to create it first. The path to the temporary folder can be either relative or absolute.</p></li>
<li><p><code>--non_bs_mm</code></p>

<p>Optionally outputs an extra column specifying the number of non-bisulfite mismatches a read during the alignment step. This option is only available for SAM format. In Bowtie 2 context, this value is just the number of actual non-bisulfite mismatches and ignores potential insertions or deletions. The format for single-end reads and read 1 of paired-end reads is <code>XA:Z:number of mismatches</code> and <code>XB:Z:number of mismatches</code> for read 2 of paired-end reads.</p></li>
<li><p><code>--gzip</code></p>

<p>Temporary bisulfite conversion files will be written out in a <code>GZIP</code> compressed form to save disk space. This option is available for most alignment modes but is not available for paired-end <code>FastA</code> files.</p></li>
<li><p><code>--sam</code></p>

<p>The output will be written out in <code>SAM</code> format instead of the default <code>BAM</code> format.</p></li>
<li><p><code>--cram</code></p>

<p>Writes the output to a <code>CRAM</code> file instead of <code>BAM</code>. This requires the use of Samtools 1.2 or higher.</p></li>
<li><p><code>--cram_ref &lt;ref_file&gt;</code></p>

<p><code>CRAM</code> output requires you to specify a reference genome as a single FastA file. If this single-FastA reference file is not supplied explicitly it will be regenerated from the genome <code>.fa</code> sequence(s) used for the Bismark run and written to a file called <code>Bismark_genome_CRAM_reference.mfa</code> into the output directory.</p></li>
<li><p><code>--samtools_path</code></p>

<p>The path to your <code>Samtools</code> installation, e.g. <code>/home/user/samtools/</code>. Does not need to be specified explicitly if <code>Samtools</code> is in the <code>PATH</code> already.</p></li>
<li><p><code>--prefix &lt;prefix&gt;</code></p>

<p>Prefixes <code>&lt;prefix&gt;</code> to the output file names. Trailing dots will be replaced by a single one. For example, <code>--prefix test</code> with <code>file.fq</code> would result in the output file <code>test.file_bismark.bam</code> etc.</p></li>
<li><p><code>-B/--basename &lt;basename&gt;</code></p>

<p>Write all output to files starting with this base file name. For example, <code>--basename foo</code> would result in the files <code>foo.bam</code> and <code>foo_SE_report.txt</code> (or its paired-end equivalent). Takes precedence over <code>--prefix</code>.</p></li>
<li><p><code>--old_flag</code></p>

<p>Only in paired-end SAM mode, uses the FLAG values used by Bismark 0.8.2 and before. In addition, this options appends /1 and /2 to the read IDs for reads 1 and 2 relative to the input file. Since both the appended read IDs and custom FLAG values may cause problems with some downstream tools such as Picard, new defaults were implemented as of version 0.8.3.</p>

<pre><code>        default                  old_flag
   ===================     ===================
   Read 1       Read 2     Read 1       Read 2
 OT:    99           147        67           131
 OB:    83           163       115           179
 CTOT:  99           147        67           131
 CTOB:  83           163       115           179
</code></pre></li>
<li><p><code>--ambig_bam</code></p>

<p>For reads that have multiple alignments a random alignment is written out to a special file ending in <code>.ambiguous.bam</code>. The alignments are in Bowtie2 format and do not any contain Bismark specific entries such as the methylation call etc. These ambiguous BAM files are intended to be used as coverage estimators for variant callers.</p></li>
<li><p><code>--nucleotide_coverage</code></p>

<p>Calculates the mono- and di-nucleotide sequence composition of covered positions in the analysed BAM file and compares it to the genomic average composition once alignments are complete by calling <code>bam2nuc</code>. Since this calculation may take a while, <code>bam2nuc</code> attempts to write the genomic sequence composition into a file called <em>genomic_nucleotide_frequencies.txt</em> inside the reference genome folder so it can be re-used the next time round instead of calculating it once again. If a file <em>nucleotide_stats.txt</em> is found with the Bismark reports it will be automatically detected and used for the Bismark HTML report. This option works only for BAM or CRAM files.</p></li>
</ul>

<h5 id="other">Other:</h5>

<ul>
<li><p><code>-h/--help</code></p>

<p>Displays this help file. Displays version information.</p></li>
<li><p><code>-v/--version</code></p>

<p>Displays version information and exits.</p></li>
</ul>

<h5 id="bowtie-specific-options">BOWTIE 2 SPECIFIC OPTIONS</h5>

<ul>
<li><p><code>--bowtie2</code> </p>

<p>Default: ON. Uses Bowtie 2. Bismark limits Bowtie 2 to only perform end-to-end alignments, i.e. searches for alignments involving all read characters (also called untrimmed or unclipped alignments). Bismark assumes that raw sequence data is adapter and/or quality trimmed where appropriate. Both small (<code>.bt2</code>) and large (<code>.bt2l</code>) Bowtie 2 indexes are supported.</p></li>
<li><p><code>--dovetail</code></p>

<p>It is possible, though unusual, for the mates to "dovetail", with the mates seemingly extending "past" each other as in this example:</p>

<pre><code>                 Mate 1:                 GTCAGCTACGATATTGTTTGGGGTGACACATTACGC
                 Mate 2:            TATGAGTCAGCTACGATATTGTTTGGGGTGACACAT
                 Reference: GCAGATTATATGAGTCAGCTACGATATTGTTTGGGGTGACACATTACGCGTCTTTGAC
</code></pre>

<p>By default, dovetailing is considered inconsistent with concordant alignment, but setting <code>--dovetail</code> causes Bowtie 2 to consider dovetailing alignments as concordant. This becomes relevant whenever reads are clipped from their 5' end prior to mapping, e.g. because of quality or bias issues. <code>--dovetail</code> is set automatically for PBAT libraries.</p></li>
</ul>

<h5 id="hisat-specific-options">HISAT2 SPECIFIC OPTIONS:</h5>

<ul>
<li><p><code>--hisat2</code> </p>

<p>Default: OFF. Uses HISAT2. Bismark limits HISAT2 to perform end-to-end alignments, i.e. searches for alignments involving all read characters (also called untrimmed or unclipped alignments) using the option <code>--no-softclipping</code>. Bismark assumes that raw sequence data is adapter and/or quality trimmed where appropriate. Both small (<code>.ht2</code>) and large (<code>.ht2l</code>) HISAT2 indexes are supported.</p></li>
<li><p><code>--no-spliced-alignment</code></p>

<p>Disable spliced alignment.</p></li>
<li><p><code>--known-splicesite-infile &lt;path&gt;</code></p>

<p>Provide a list of known splice sites.</p></li>
</ul>

<h5 id="paired-end-options">Paired-end options:</h5>

<ul>
<li><p><code>--no-mixed</code></p>

<p>This option disables the behaviour to try to find alignments for the individual mates if it cannot find a concordant or discordant alignment for a pair. This option is invariably on by default.</p></li>
<li><p><code>--no-discordant</code></p>

<p>Normally, Bowtie 2 or HISAT2 look for discordant alignments if they cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints (<code>--fr</code>/<code>--rf</code>/<code>--ff</code>, <code>-I</code>, <code>-X</code>). This option disables that behaviour and is on by default.</p></li>
</ul>

<h5 id="bowtie-effort-options">Bowtie 2 Effort options:</h5>

<ul>
<li><p><code>-D &lt;int&gt;</code></p>

<p>Up to &lt;int&gt; consecutive seed extension attempts can "fail" before Bowtie 2 moves on, using the alignments found so far. A seed extension "fails" if it does not yield a new best or a new second-best alignment. Default: 15.</p></li>
<li><p><code>-R &lt;int&gt;</code> </p>

<p>&lt;int&gt; is the maximum number of times Bowtie 2 will "re-seed" reads with repetitive seeds. When "re-seeding," Bowtie 2 simply chooses a new set of reads (same length, same number of mismatches allowed) at different offsets and searches for more alignments. A read is considered to have repetitive seeds if the total number of seed hits divided by the number of seeds that aligned at least once is greater than 300. Default: 2.</p></li>
</ul>

<h5 id="parallelization-options">Parallelization options:</h5>

<ul>
<li><p><code>-p NTHREADS</code></p>

<p>Launch NTHREADS parallel search threads (default: 1). Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speed-up is close to linear. NOTE: It is currently unclear whether this speed increase also translates into a speed increase of Bismark since it is running several instances of Bowtie 2 concurrently! Increasing <code>-p</code> increases Bowtie 2's memory footprint. E.g. when aligning to a human genome index, increasing <code>-p</code> from 1 to 8 increases the memory footprint by a few hundred megabytes (for each instance of Bowtie 2!). This option is only available if Bowtie is linked with the pthreads library (i.e. if BOWTIE_PTHREADS=0 is not specified at build time). In addition, this option will automatically use the option <code>--reorder</code>, which guarantees that output SAM records are printed in an order corresponding to the order of the reads in the original input file, even when <code>-p</code> is set greater than 1 (Bismark requires the Bowtie 2 output to be this way). Specifying <code>--reorder</code> and setting <code>-p</code> greater than 1 causes Bowtie 2 to run somewhat slower and use somewhat more memory than if <code>--reorder</code> were not specified. Has no effect if <code>-p</code> is set to 1, since output order will naturally correspond to input order in that case.</p></li>
</ul>

<h5 id="scoring-options">Scoring options:</h5>

<ul>
<li><p><code>--score_min &lt;func&gt;</code></p>

<p>Sets a function governing the minimum alignment score needed for an alignment to be considered "valid" (i.e. good enough to report). This is a function of read length. For instance, specifying <code>L,0,-0.2</code> sets the minimum-score function <code>f</code> to <code>f(x) = 0 + -0.2 * x</code>, where <code>x</code> is the read length. See also: setting function options at http://bowtie-bio.sourceforge.net/bowtie2. The default is: L,0,-0.2.</p></li>
<li><p><code>--rdg &lt;int1&gt;,&lt;int2&gt;</code> </p>

<p>Sets the read gap open (&lt;int1>) and extend (&lt;int2>) penalties. A read gap of length N gets a penalty of <code>&lt;int1&gt; + N * &lt;int2&gt;</code>. Default: 5, 3.</p></li>
<li><p><code>--rfg &lt;int1&gt;,&lt;int2&gt;</code></p>

<p>Sets the reference gap open (&lt;int1>) and extend (&lt;int2>) penalties. A reference gap of length N gets a penalty of <code>&lt;int1&gt; + N * &lt;int2&gt;</code>. Default: 5, 3.</p></li>
</ul>

<h2 id="appendix-iii-bismark-methylation-extractor">Appendix (III): Bismark Methylation Extractor</h2>

<p>A brief description of the Bismark methylation extractor and a full list of options can also be viewed by typing <code>bismark_methylation_extractor --help</code></p>

<h4 id="usage-bismark-methylation-extractor-options-lt-filenames-gt">USAGE: <code>bismark_methylation_extractor [options] &lt;filenames&gt;</code></h4>

<h4 id="arguments">ARGUMENTS:</h4>

<ul>
<li><p><code>&lt;filenames&gt;</code></p>

<p>A space-separated list of Bismark result files in SAM format from which methylation information is extracted for every cytosine in the reads. For alignment files in the older custom Bismark output see option <code>--vanilla</code>.</p></li>
</ul>

<h4 id="options">OPTIONS:</h4>

<ul>
<li><p><code>-s/--single-end</code></p>

<p>Input file(s) are Bismark result file(s) generated from single-end read data. If neither <code>-s</code> nor <code>-p</code> is set the type of experiment will be determined automatically.</p></li>
<li><p><code>-p/--paired-end</code></p>

<p>Input file(s) are Bismark result file(s) generated from paired-end read data. If neither <code>-s</code> nor <code>-p</code> is set the type of experiment will be determined automatically.</p></li>
<li><p><code>--vanilla</code></p>

<p>The Bismark result input file(s) are in the old custom Bismark format (up to version 0.5.x) and not in SAM format which is the default as of Bismark version 0.6.x or higher. Default: OFF.</p></li>
<li><p><code>--no_overlap</code></p>

<p>For paired-end reads it is theoretically possible that Read 1 and Read 2 overlap. This option avoids scoring overlapping methylation calls twice (only methylation calls of read 1 are used for in the process since read 1 has historically higher quality basecalls than read 2). Whilst this option removes a bias towards more  methylation calls in the center of sequenced fragments it may <em>de facto</em> remove a sizeable proportion of the data. This option is on by default for paired-end data but can be disabled using <code>--include_overlap</code>.  Default: ON.</p></li>
<li><p><code>--include_overlap</code></p>

<p>For paired-end data all methylation calls will be extracted irrespective of whether they overlap or not. Default: OFF.</p></li>
<li><p><code>--ignore &lt;int&gt;</code></p>

<p>Ignore the first &lt;int> bp from the 5' end of Read 1 (or single-end alignment files) when processing the methylation call string. This can remove e.g. a restriction enzyme site at the start of each read or any other source of bias (such as PBAT-Seq data).</p></li>
<li><p><code>--ignore_r2 &lt;int&gt;</code></p>

<p>Ignore the first &lt;int> bp from the 5' end of Read 2 of paired-end sequencing results only. Since the first couple of bases in Read 2 of BS-Seq experiments show a severe bias towards non-methylation as a result of end-repairing sonicated fragments with unmethylated cytosines (see M-bias plot), it is recommended that the first couple of bp of Read 2 are removed before starting downstream analysis. Please see the section on M-bias plots in the Bismark User Guide for more details.</p></li>
<li><p><code>--ignore_3prime &lt;int&gt;</code></p>

<p>Ignore the last &lt;int> bp from the 3' end of Read 1 (or single-end alignment files) when processing the methylation call string. This can remove unwanted biases from the end of reads.</p></li>
<li><p><code>--ignore_3prime_r2 &lt;int&gt;</code> </p>

<p>Ignore the last &lt;int> bp from the 3' end of Read 2 of paired-end sequencing results only. This can remove unwanted biases from the end of reads.</p></li>
<li><p><code>--comprehensive</code></p>

<p>Specifying this option will merge all four possible strand-specific methylation info into context-dependent output files. The default contexts are:</p>

<ul>
<li>CpG context
</li>
<li>CHG context
</li>
<li>CHH context
</li>
</ul></li>
<li><p><code>--merge_non_CpG</code></p>

<p>This will produce two output files (in <code>--comprehensive mode</code>) or eight strand-specific output files (default) for Cs in </p>

<ul>
<li>CpG context
</li>
<li>non-CpG context
</li>
</ul></li>
<li><p><code>--report</code></p>

<p>Prints out a short methylation summary as well as the parameters used to run this script. Default: ON.</p></li>
<li><p><code>--no_header</code></p>

<p>Suppresses the Bismark version header line in all output files for more convenient batch processing.</p></li>
<li><p><code>-o/--output DIR</code></p>

<p>Allows specification of a different output directory (absolute or relative path). If not specified explicitly, the output will be written to the current directory.</p></li>
<li><p><code>--samtools_path</code></p>

<p>The path to your Samtools installation, e.g. /home/user/samtools/. Does not need to be specified explicitly if Samtools is in the PATH already.</p></li>
<li><p><code>--gzip</code></p>

<p>The methylation extractor files (CpG_OT_..., CpG_OB_... etc) will be written out in a <code>GZIP</code> compressed form to save disk space. This option is also passed on to the genome-wide cytosine report. <code>bedGraph</code> and <code>coverage</code> files are written out as <code>.gz</code> by default.</p></li>
<li><p><code>--mbias_only</code></p>

<p>The methylation extractor will read the entire file but only output the M-bias table and plots as well as a report (optional) and then quit. Default: OFF.</p></li>
<li><p><code>--mbias_off</code></p>

<p>The methylation extractor will process the entire file as usual but doesn't write out any M-bias report. Only recommended for users who deliberately want to keep an earlier version of the M-bias report. Default: OFF.</p></li>
<li><p><code>--multicore &lt;int&gt;</code></p>

<p>Sets the number of cores to be used for the methylation extraction process. If system resources are plentiful this is a viable option to speed up the extraction process (we observed a near linear speed increase for up to 10 cores used). Please note that a typical process of extracting a BAM file and writing out <code>.gz</code> output streams will in fact use ~3 cores per value of <code>--multicore &lt;int&gt;</code> specified (1 for the methylation extractor itself, 1 for a Samtools stream, 1 for a GZIP stream), so <code>--multicore 10</code> is likely to use around 30 cores of system resources. This option has no bearing on the <code>bismark2bedGraph</code> or <code>coverage2cytosine</code> (genome-wide cytosine report) processes.</p></li>
<li><p><code>--version</code></p>

<p>Displays version information.</p></li>
<li><p><code>-h/--help</code></p>

<p>Displays this help file and exits.</p></li>
</ul>

<h5 id="bedgraph-specific-options">bedGraph specific options:</h5>

<ul>
<li><p><code>--bedGraph</code></p>

<p>After finishing the methylation extraction, the methylation output is written into a sorted <code>bedGraph</code> file that reports the position of a given cytosine and its methylation state (in %, see details below) using 0-based genomic start and 1-based end coordinates. The methylation extractor output is temporarily split up into temporary files, one per chromosome (written into the current directory or folder specified with <code>-o/--output</code>); these temp files are then used for sorting and deleted afterwards. By default, only cytosines in CpG context are sorted. The option <code>--CX_context</code> may be used to report all cytosines irrespective of sequence context (this will take <em>MUCH</em> longer!). The <code>bedGraph</code> conversion step is performed by the external module <code>bismark2bedGraph</code>; this script needs to reside in the same folder as the bismark_methylation_extractor itself.</p></li>
<li><p><code>--zero_based</code></p>

<p>Write out an additional coverage file (ending in <code>.zero.cov</code>) that uses 0-based genomic start and 1-based genomic end coordinates (zero-based, half-open), like used in the <code>bedGraph</code> file, instead of using 1-based coordinates throughout. Default: OFF.</p></li>
<li><p><code>--cutoff [threshold]</code></p>

<p>The minimum number of times any methylation state (methylated or unmethylated) has to be seen for a nucleotide before its methylation percentage is reported. Default: 1.</p></li>
<li><p><code>--remove_spaces</code></p>

<p>Replaces white spaces in the sequence ID field with underscores to allow sorting.</p></li>
<li><p><code>--CX/--CX_context</code></p>

<p>The sorted <code>bedGraph</code> output file contains information on every single cytosine that was covered in the experiment irrespective of its sequence context. This applies to both forward and reverse strands. Please be aware that this option may generate large temporary and output files and may take a long time to sort (up to many hours). Default: OFF (i.e. Default = CpG context only).</p></li>
<li><p><code>--buffer_size &lt;string&gt;</code></p>

<p>This allows you to specify the main memory sort buffer when sorting the methylation information. Either specify a percentage of physical memory by appending % (e.g. <code>--buffer_size 50%</code>) or a multiple of 1024 bytes, e.g. <code>K</code> multiplies by 1024, <code>M</code> by 1048576 and so on for <code>T</code> etc. (e.g. <code>--buffer_size 20G</code>). For more information on sort type <code>info sort</code> on a command line. Defaults to 2G.</p></li>
<li><p><code>--scaffolds/--gazillion</code> </p>

<p>Users working with unfinished genomes sporting tens or even hundreds of thousands of scaffolds/contigs/chromosomes frequently encountered errors with pre-sorting reads to individual chromosome files. These errors were caused by the operating system's limit of the number of filehandles that can be written to at any one time (typically 1024; to find out this limit on Linux, type: <code>ulimit -a</code>). To bypass the limitation of open filehandles, the option <code>--scaffolds</code> does not pre-sort methylation calls into individual chromosome files. Instead, all input files are temporarily merged into a single file (unless there is only a single file), and this file will then be sorted by both chromosome AND position using the Unix sort command. Please be aware that this option might take a l*ooooo*ng time to complete, depending on the size of the input files, and the memory you allocate to this process (see <code>--buffer_size</code>). Nevertheless, it seems to be working.</p></li>
<li><p><code>--ample_memory</code></p>

<p>Using this option will not sort chromosomal positions using the UNIX <code>sort</code> command, but will instead use two arrays to sort methylated and unmethylated calls. This may result in a faster sorting process of very large files, but this comes at the cost of a larger memory footprint (two arrays of the length of the largest human chromosome 1 (~250M bp) consume around 16GB of RAM). Due to overheads in creating and looping through these arrays it seems that it will actually be <em>slower</em> for small files (few million alignments), and we are currently testing at which point it is advisable to use this option. Note that <code>--ample_memory</code> is not compatible with options <code>--scaffolds/--gazillion</code> (as it requires pre-sorted files to begin with).</p></li>
</ul>

<h5 id="genome-wide-cytosine-methylation-report-specific-options">Genome-wide cytosine methylation report specific options:</h5>

<ul>
<li><p><code>--cytosine_report</code></p>

<p>After the conversion to bedGraph has completed, the option <code>--cytosine_report</code> produces a genome-wide methylation report for all cytosines in the genome. By default, the output uses 1-based chromosome coordinates (zero-based start coords are optional) and reports CpG context only (all cytosine context is optional). The output considers all Cs on both forward and reverse strands and reports their position, strand, trinucleotide content and methylation state (counts are 0 if not covered). The cytosine report conversion step is performed by the external module <code>coverage2cytosine</code>; this script needs to reside in the same folder as the bismark_methylation_extractor itself.</p></li>
<li><p><code>--CX/--CX_context</code></p>

<p>The output file contains information on every single cytosine in the genome irrespective of its context. This applies to both forward and reverse strands. Please be aware that this will generate output files with > 1.1 billion lines for a mammalian genome such as human or mouse. Default: OFF (i.e. Default = CpG context only).</p></li>
<li><p><code>--zero_based</code></p>

<p>Uses 0-based genomic coordinates instead of 1-based coordinates. Default: OFF.</p></li>
<li><p><code>--genome_folder &lt;path&gt;</code></p>

<p>Enter the genome folder you wish to use to extract sequences from (full path only). Accepted formats are FastA files ending with <code>.fa</code> or <code>.fasta</code>. Specifying a genome folder path is mandatory.</p></li>
<li><p><code>--split_by_chromosome</code></p>

<p>Writes the output into individual files for each chromosome instead of a single output file. Files are named to include the input filename as well as the chromosome number.</p></li>
</ul>

<h4 id="output">OUTPUT</h4>

<h5 id="the-bismark-methylation-extractor-output-is-in-the-form-tab-delimited-based-coords">The bismark_methylation_extractor output is in the form (tab delimited, 1-based coords):</h5>

<pre><code>&lt;seq-ID&gt; &lt;methylation state*&gt; &lt;chromosome&gt; &lt;start position (= end position)&gt; &lt;methylation call&gt;

  Methylated cytosines receive a '+' orientation,
Unmethylated cytosines receive a '-' orientation.
</code></pre>

<h5 id="the-bedgraph-output-optional-looks-like-this-tab-delimited-based-start-based-end-coords">The bedGraph output (optional) looks like this (tab-delimited, 0-based start, 1-based end coords):</h5>

<pre><code>track type=bedGraph (header line)
&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt;
</code></pre>

<h5 id="the-coverage-output-looks-like-this-tab-delimited-based-genomic-coords">The coverage output looks like this (tab-delimited; 1-based genomic coords):</h5>

<pre><code>&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt; &lt;count methylated&gt; &lt;count unmethylated&gt;
</code></pre>

<h5 id="the-genome-wide-cytosine-report-optional-is-tab-delimited-in-the-following-format-based-coords">The genome-wide cytosine report (optional) is tab-delimited in the following format (1-based coords):</h5>

<pre><code>&lt;chromosome&gt; &lt;position&gt; &lt;strand&gt; &lt;count methylated&gt; &lt;count unmethylated&gt; &lt;C-context&gt; &lt;trinucleotide context&gt;
</code></pre>

<h2 id="appendix-iv-bismark-reports-for-the-test-data-set">Appendix (IV): Bismark reports for the test data set</h2>

<p>Please note that this has been run with a fairly early version however I wouldn't expect the numbers to change much.</p>

<h5 id="using-bowtie">Using Bowtie 2:</h5>

<pre><code>Running Bismark with the following options (e.g. bismark --score-min L,0,-0.6 /data/public/Genomes/Human/GRCh37/ test_data.fastq) should result in this mapping report:

Bismark report for: test_data.fastq (version: v0.7.8)
Option '--directional' specified: alignments to complementary strands will be ignored (i.e. not performed!)
Bowtie2 was run against the bisulfite genome of /data/public/Genomes/Human/GRCh37/ with the specified options: -q -- score-min L,0,-0.6 --ignore-quals

Final Alignment report
======================
Sequences analysed in total: 10000

Number of alignments with a unique best hit from the different alignments: 5658 Mapping efficiency: 56.6%
Sequences with no alignments under any condition: 2893
Sequences did not map uniquely: 1449
Sequences which were discarded because genomic sequence could not be extracted: 0
Number of alignments to (merely theoretical) complementary strands being rejected in total: 0

Number of sequences with unique best (first) alignment came from the bowtie output:

CT/CT: 2820 ((converted) top strand) 
CT/GA: 2838 ((converted) bottom strand)
GA/CT: 0    (complementary to (converted) top strand) 
GA/GA: 0    (complementary to (converted) bottom strand)

Final Cytosine Methylation Report
================================= 
Total number of C's analysed: 45985

Total methylated C's in CpG context: 1550
Total methylated C's in CHG context: 34 
Total methylated C's in CHH context: 126

Total C to T conversions in CpG context: 844
Total C to T conversions in CHG context: 11368
Total C to T conversions in CHH context:32063

C methylated in CpG context: 64.7% 
C methylated in CHG context: 0.3% 
C methylated in CHH context: 0.4%
</code></pre>

<h1 id="credits">Credits</h1>

<p>Bismark was written by Felix Krueger at the <a href="http://www.bioinformatics.babraham.ac.uk/">Babraham Bioinformatics Group</a>.</p>

<p><img src="Images/bioinformatics_logo.png" alt="Babraham Bioinformatics" /></p>

</body>
</html>
